{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6169c8c4-6c46-41c1-9c84-3bff9d2e2c8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sbol2\n",
    "import os\n",
    "import requests\n",
    "from dotenv import load_dotenv\n",
    "import zipfile\n",
    "from __future__ import absolute_import, division, print_function\n",
    "import numpy as np\n",
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import tempfile\n",
    "from abc import abstractmethod, ABCMeta\n",
    "from sklearn.tree import DecisionTreeClassifier as scikit_DecisionTree\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.preprocessing import OneHotEncoder, StandardScaler\n",
    "from sklearn.model_selection import train_test_split, GridSearchCV\n",
    "import sys\n",
    "import argparse\n",
    "import time\n",
    "from sklearn.metrics import mean_squared_error, r2_score, accuracy_score\n",
    "import xgboost as xgb\n",
    "from sklearn.metrics import mean_squared_error\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "current_dir = os.path.abspath('')\n",
    "data_path = os.path.join(current_dir, '..', 'data')\n",
    "attachments_path = os.path.join(current_dir, '..', 'attachments')\n",
    "pulled_attachments_path = os.path.join(current_dir, '..', 'pulled_attachments')\n",
    "sbol_path = os.path.join(current_dir, '..', 'sbol_data')\n",
    "downloaded_sbol_path = os.path.join(current_dir, '..', 'downloaded_sbol')\n",
    "original_data_path = os.path.join(data_path, 'original_data')\n",
    "scripts_path = os.path.join(current_dir, 'scripts')\n",
    "model_data_path = os.path.join(data_path, 'processed_data', 'replicated_models')\n",
    "model_output_path = os.path.join('..', 'model_outputs')\n",
    "\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "05b881c7-b711-41ab-af99-5c3e0dcb7245",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(os.path.join(original_data_path, \"frag-rLP5_LB_expression.txt\"), delimiter=\" \")\n",
    "# Expression levels from genomic fragment MPRA (random 200–300 bp sheared fragments) in LB media\t\n",
    "df2 = pd.read_csv(os.path.join(original_data_path, \"frag-rLP5-M9_expression.txt\"), delimiter=\" \")\n",
    "# Expression levels from genomic fragment MPRA (random 200–300 bp sheared fragments) in M9 media at rLP5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "761f55c1-a68b-485f-826f-09a11af50b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_sbol_doc(df, file_path, media=\"LB\"):\n",
    "    doc = sbol2.Document()\n",
    "    sbol2.setHomespace('http://github.com/cywlol/promoters')\n",
    "    doc.displayId = \"E_coli_promoters\"\n",
    "    # save labels for later when we do the attachments through api\n",
    "    exp_data_labels = []\n",
    "    attachment_file_names = []\n",
    "\n",
    "    media_label_MD = media\n",
    "    chassis_label_MD = \"E_coli_chassis\"\n",
    "\n",
    "    # define the media once, then have it as a reference for each promoter row\n",
    "    media_md = sbol2.ModuleDefinition(media_label_MD)\n",
    "    doc.addModuleDefinition(media_md)\n",
    "    media_md.addRole(\"http://identifiers.org/ncit/NCIT:C48164\") \n",
    "    \n",
    "    # define the chassis once, then have it as a reference for each promoter row\n",
    "    chassis_md = sbol2.ModuleDefinition(chassis_label_MD) \n",
    "    doc.addModuleDefinition(chassis_md)\n",
    "    chassis_md.addRole(\"http://identifiers.org/ncit/NCIT:C14419\")\n",
    "    \n",
    "    #attach genome \n",
    "    chassis_md.wasDerivedFrom = [\"https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=511145\"]\n",
    "\n",
    "    for i, row in df.iterrows():\n",
    "        fragment_seq = row[\"fragment\"]\n",
    "\n",
    "        # Establish the identities of each component, ensuring that each one is unique\n",
    "        promoter_label_CD = f\"promoter_{i}\"\n",
    "        promoter_seq_label = f\"promoter_seq_{i}\"\n",
    "        sample_design_label_CD = f\"sample_design_{i}\"\n",
    "        strain_label_MD = f\"strain_{i}\"\n",
    "        data_label = f\"exp_data_{i}\"\n",
    "        exp_label = f'sample_{i}_expression_data'\n",
    "        attachment_file_name = f'frag_{media}_exp_sample_{i}.csv'\n",
    "\n",
    "        #engr_region_label_CD = f\"engr_region_{i}\"\n",
    "        #location_promoter_annotation = f\"location_promoter_annotation_{i}\"\n",
    "        #location_promoter_label = f\"location_promoter_label_{i}\"\n",
    "        \n",
    "        # Promoter and Sequence \n",
    "        promoter_cd = sbol2.ComponentDefinition(promoter_label_CD, sbol2.BIOPAX_DNA)\n",
    "        promoter_cd.roles = [sbol2.SO_PROMOTER]\n",
    "        seq = sbol2.Sequence(promoter_seq_label, fragment_seq, sbol2.SBOL_ENCODING_IUPAC)\n",
    "        doc.addSequence(seq)\n",
    "        promoter_cd.sequences = [seq.persistentIdentity]\n",
    "        doc.addComponentDefinition(promoter_cd)\n",
    "\n",
    "        \n",
    "        # Add strain module definition\n",
    "        strain_md = sbol2.ModuleDefinition(strain_label_MD)\n",
    "        strain_md.addRole(\"http://identifiers.org/ncit/NCIT:C14419\")\n",
    "        doc.addModuleDefinition(strain_md)\n",
    "        strain_c1 = strain_md.modules.create('chassis')\n",
    "        strain_c1.definition = chassis_md.identity\n",
    "        strain_c2 = strain_md.functionalComponents.create('promoter')\n",
    "        strain_c2.definition = promoter_cd.identity\n",
    "    \n",
    "        #annotation = sbol2.SequenceAnnotation(\"promoter_location\")\n",
    "        #range = sbol2.Range(\"prange\", start, end)\n",
    "    \n",
    "        #if (strand == \"-\"):\n",
    "        #    range.orientation = sbol2.SBOL_ORIENTATION_REVERSE_COMPLEMENT\n",
    "                \n",
    "        #annotation.locations.add(range)\n",
    "        #promoter_cd.sequenceAnnotations.add(annotation)\n",
    "        '''\n",
    "        # Engineered Region \n",
    "        engineered_cd = sbol2.ComponentDefinition(engr_region_label_CD, sbol2.BIOPAX_DNA)\n",
    "        engineered_cd.roles = [\"https://identifiers.org/so/SO:0000804\"]\n",
    "        sub = engineered_cd.components.create('promoter')\n",
    "        sub.definition = promoter_cd.persistentIdentity\n",
    "        doc.addComponentDefinition(engineered_cd)\n",
    "        # No sequence?\n",
    "        '''\n",
    "        \n",
    "        #strain_c2 = strain_md.functionalComponents.create('engineered_region')\n",
    "        #strain_c2.definition = strain_md.persistentIdentity\n",
    "        \n",
    "        # Sample Design  \n",
    "        sample_md = sbol2.ModuleDefinition(sample_design_label_CD)\n",
    "        doc.addModuleDefinition(sample_md)\n",
    "        sample_md.addRole(\"http://identifiers.org/obo/OBI:0000073\")\n",
    "\n",
    "        m_strain = sample_md.modules.create('strain')\n",
    "        m_strain.definition = strain_md.identity\n",
    "        \n",
    "        m_media = sample_md.modules.create('media')\n",
    "        m_media.definition = media_md.identity\n",
    "\n",
    "        # Create the attachment data as a csv\n",
    "        rna_series = row[[\"RNA_exp_1\", \"RNA_exp_2\", \"RNA_exp_ave\"]]\n",
    "        rna_df = pd.DataFrame([rna_series])  \n",
    "        rna_df.to_csv(os.path.join(attachments_path, attachment_file_name), index=False)\n",
    "        \n",
    "        # exp_attachment = sbol2.Attachment(exp_label)\n",
    "        # exp_attachment.name = f'fragmentation expression at rLP5 for sample {i}'\n",
    "        # exp_attachment.description = 'CSV including the gene expression of the sequence: RNA_exp1, RNA_exp2, and its average.'\n",
    "        # exp_attachment.source = 'CSV_LINK_HERE' # update when added attachment to SBOL collection\n",
    "        # exp_attachment.format = 'https://identifiers.org/edam/format_3752'\n",
    "        # doc.addAttachment(exp_attachment)\n",
    "        \n",
    "        # Experiment and Measurement Data\n",
    "        exp = sbol2.ExperimentalData(exp_label)\n",
    "        exp.wasDerivedFrom = sample_md.identity\n",
    "        doc.add(exp)\n",
    "\n",
    "        exp_data_labels.append(exp_label)\n",
    "        attachment_file_names.append(attachment_file_name)\n",
    "\n",
    "        if (i == 5):\n",
    "            break\n",
    "            \n",
    "    report = doc.validate()\n",
    "    if (report == 'Valid.'):\n",
    "        doc.write(file_path)\n",
    "    else:\n",
    "        print(report)\n",
    "    return exp_data_labels, attachment_file_names, doc, chassis_label_MD\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "def partshop_attach_exp_data(synbio_username, collection_name, file_names, email, password, exp_data_labels, version=1):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    print(shop.login(email, password))\n",
    "    exp_labels = [\"ExperimentalData_\" + label for label in exp_data_labels]  \n",
    "    for label, file_name in zip(exp_labels, file_names):\n",
    "        path = os.path.join(attachments_path, file_name)\n",
    "        attachment_uri = f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{label}/{version}\"\n",
    "        shop.attachFile(attachment_uri, path)\n",
    "\n",
    "def partshop_attach_genome_to_md(synbio_username, collection_name, file_path, email, password, chassis_label, version=1):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    print(shop.login(email, password))\n",
    "\n",
    "    label = \"ModuleDefinition_\" + chassis_label\n",
    "    attachment_uri = f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{label}/{version}\"\n",
    "\n",
    "    shop.attachFile(attachment_uri, file_path)\n",
    "\n",
    "def create_synbio_collection(email, password, file_path, id, name, description, version='1'):\n",
    "    response = requests.post(\n",
    "        \"https://synbiohub.org/login\",\n",
    "        headers={\"Accept\": \"text/plain\"},\n",
    "        data={\"email\": email, \"password\": password}\n",
    "    )\n",
    "        \n",
    "    if response.ok:\n",
    "        token = response.text.strip() # theres a whitespace before the token for some reason\n",
    "        response = requests.post(\n",
    "        'https://synbiohub.org/submit',\n",
    "        headers={\n",
    "            'X-authorization': token,\n",
    "            'Accept': 'text/plain'\n",
    "        },\n",
    "        files={\n",
    "        'files': open(file_path,'rb'),\n",
    "        },\n",
    "        data={\n",
    "            'id': id,\n",
    "            'version' : version,\n",
    "            'name' :  name,\n",
    "            'description' : description,\n",
    "            'citations' : '',\n",
    "            'overwrite_merge' : '0'\n",
    "        },\n",
    "    \n",
    "    )\n",
    "    else:\n",
    "        print(\"Login failed:\", response.status_code)\n",
    "        print(response.text)\n",
    "\n",
    "def partshop_pull(email, password, synbio_username, collection_name, file_path, version=1):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    doc = sbol2.Document()\n",
    "    shop.login(email, password)\n",
    "\n",
    "    collection_uri = f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{collection_name}_collection/{version}\"\n",
    "    s = shop.pull(collection_uri, doc)\n",
    "    \n",
    "    for obj in doc:\n",
    "        print(obj)   \n",
    "    \n",
    "    doc.write(file_path)\n",
    "\n",
    "def download_all_attachments(email, password, doc, file_path):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    shop.login(email, password)\n",
    "    for attachment in doc.attachments:\n",
    "        shop.downloadAttachment(attachment.identity, filepath=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b4e5cf9e-6ff7-4261-8461-a11f6397f205",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sbol2\n",
    "\n",
    "def create_sbol_attempt2_doc(df, file_path, media=\"LB\"):\n",
    "    doc = sbol2.Document()\n",
    "    sbol2.setHomespace('http://github.com/cywlol/promoters')\n",
    "    doc.displayId = \"E_coli_promoters\"\n",
    "\n",
    "    media_label_MD = media\n",
    "    chassis_label_MD = \"E_coli_chassis\"\n",
    "\n",
    "\n",
    "    # define the media once, then have it as a reference for each promoter row\n",
    "    media_md = sbol2.ModuleDefinition(media_label_MD)\n",
    "    media_md.addRole(\"http://identifiers.org/ncit/NCIT:C48164\")\n",
    "    media_md.version = \"1\"\n",
    "    doc.addModuleDefinition(media_md)\n",
    "\n",
    "    # define the chassis once, then have it as a reference for each promoter row\n",
    "    chassis_md = sbol2.ModuleDefinition(chassis_label_MD)\n",
    "    chassis_md.addRole(\"http://identifiers.org/ncit/NCIT:C14419\")\n",
    "    chassis_md.wasDerivedFrom = [\"https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=511145\"]\n",
    "    chassis_md.version = \"1\"\n",
    "    doc.addModuleDefinition(chassis_md)\n",
    "\n",
    "    for i, row in df.iterrows():\n",
    "        fragment_seq = row[\"fragment\"]\n",
    "        start = row[\"start\"]\n",
    "        end = row[\"end\"]\n",
    "        strand = row[\"strand\"]\n",
    "        RNA_exp_1 = row[\"RNA_exp_1\"]\n",
    "        RNA_exp_2 = row[\"RNA_exp_2\"]\n",
    "        RNA_exp_ave = row[\"RNA_exp_ave\"]\n",
    "\n",
    "        # Establish the identities of each component, ensuring that each one is unique\n",
    "        promoter_label_CD = f\"promoter_definition_{i}\"\n",
    "        promoter_label_FC = f\"promoter_fc_{i}\"\n",
    "        promoter_seq_label = f\"promoter_seq_{i}\"\n",
    "        sample_design_label_CD = f\"sample_design_definition_{i}\"\n",
    "        chassis_module_label = f\"chassis_module_{i}\"\n",
    "        circuit_label_CD = f\"circuit_definition_{i}\"\n",
    "        strain_label_MD = f\"strain_definition_{i}\"\n",
    "        circuit_label_FC = f\"cicuit_fc_{i}\"\n",
    "        measurement_label = f\"exp_ave_{i}\"\n",
    "        interaction_label = f\"interaction_{i}\"\n",
    "        gfp_fc_label = f\"gfp_{i}\"\n",
    "        promoter_location_label = f\"promoter_{i}_location\"\n",
    "        strain_module_label = f\"strain_module_{i}\"\n",
    "        media_module_label = f\"media_module_{i}\"\n",
    "        # define gfp once\n",
    "        gfp_component_label = f\"gfp_{i}\"\n",
    "\n",
    "\n",
    "\n",
    "        #engr_region_label_CD = f\"engr_region_{i}\"\n",
    "        #location_promoter_annotation = f\"location_promoter_annotation_{i}\"\n",
    "        #location_promoter_label = f\"location_promoter_label_{i}\"\n",
    "\n",
    "\n",
    "        '''\n",
    "        # Engineered Region\n",
    "        engineered_cd = sbol2.ComponentDefinition(engr_region_label_CD, sbol2.BIOPAX_DNA)\n",
    "        engineered_cd.roles = [\"https://identifiers.org/so/SO:0000804\"]\n",
    "        sub = engineered_cd.components.create('promoter')\n",
    "        sub.definition = promoter_cd.persistentIdentity\n",
    "        doc.addComponentDefinition(engineered_cd)\n",
    "        # No sequence?\n",
    "        '''\n",
    "\n",
    "\n",
    "        # Add strain module definition\n",
    "        strain_md = sbol2.ModuleDefinition(strain_label_MD)\n",
    "        strain_md.addRole(\"http://identifiers.org/ncit/NCIT:C14419\")\n",
    "        strain_md.version = \"1\"\n",
    "        doc.addModuleDefinition(strain_md)\n",
    "\n",
    "        chassis_module = sbol2.Module(chassis_module_label)\n",
    "        chassis_module.definition = chassis_md.identity\n",
    "        strain_md.modules.add(chassis_module)\n",
    "\n",
    "        # Sample Design\n",
    "        sample_md = sbol2.ModuleDefinition(sample_design_label_CD)\n",
    "        sample_md.addRole(\"http://identifiers.org/obo/OBI:0000073\")\n",
    "        sample_md.version = \"1\"\n",
    "        doc.addModuleDefinition(sample_md)\n",
    "\n",
    "        # Add strain module to sample design\n",
    "        strain_module = sbol2.Module(strain_module_label)\n",
    "        strain_module.definition = strain_md.identity\n",
    "        # Add media module to sample design\n",
    "        media_module = sbol2.Module(media_module_label)\n",
    "        media_module.definition = media_md.identity\n",
    "\n",
    "\n",
    "        sample_md.modules.add(media_module)\n",
    "        sample_md.modules.add(strain_module)\n",
    "\n",
    "        # Promoter and Sequence\n",
    "        promoter_cd = sbol2.ComponentDefinition(promoter_label_CD, sbol2.BIOPAX_DNA)\n",
    "        promoter_cd.roles = [sbol2.SO_PROMOTER]\n",
    "        promoter_cd.version = \"1\"\n",
    "        doc.addComponentDefinition(promoter_cd)\n",
    "        seq = sbol2.Sequence(promoter_seq_label, fragment_seq, sbol2.SBOL_ENCODING_IUPAC)\n",
    "        seq.version = \"1\"\n",
    "\n",
    "        doc.addSequence(seq)\n",
    "        promoter_cd.sequences = seq\n",
    "\n",
    "        # Circuit\n",
    "        circuit = sbol2.ComponentDefinition(circuit_label_CD, sbol2.BIOPAX_DNA)\n",
    "        circuit.roles = [\"https://identifiers.org/so/SO:0000804\"]\n",
    "        circuit.version = \"1\"\n",
    "\n",
    "        annotation = sbol2.SequenceAnnotation(promoter_location_label)\n",
    "        annotation.version = \"1\"\n",
    "        rng = sbol2.Range(\"prange\", start, end)\n",
    "        rng.version = \"1\"\n",
    "\n",
    "        if (strand == \"-\"):\n",
    "            rng.orientation = sbol2.SBOL_ORIENTATION_REVERSE_COMPLEMENT\n",
    "\n",
    "        annotation.locations.add(rng)\n",
    "        circuit.sequenceAnnotations.add(annotation)\n",
    "        promoter_comp = circuit.components.create('promoter')\n",
    "        promoter_comp.definition = promoter_cd.identity\n",
    "        doc.addComponentDefinition(circuit)\n",
    "\n",
    "\n",
    "        circuit_fc = sbol2.FunctionalComponent(circuit_label_FC)\n",
    "        circuit_fc.version = \"1\"\n",
    "        circuit_fc.definition = circuit.identity\n",
    "        circuit_fc.direction = sbol2.SBOL_DIRECTION_OUT\n",
    "        circuit_fc.access = sbol2.SBOL_ACCESS_PUBLIC\n",
    "        strain_md.functionalComponents.add(circuit_fc)\n",
    "\n",
    "        measurement3 = sbol2.Measurement(measurement_label, value=RNA_exp_ave, unit=\"http://www.ontology-of-units-of-measure.org/resource/om-2/RatioUnit\")\n",
    "        measurement3.version = \"1\"\n",
    "\n",
    "        gfp_protein = sbol2.ComponentDefinition(gfp_component_label, sbol2.BIOPAX_PROTEIN)\n",
    "        gfp_protein.roles=[sbol2.SO_CDS]\n",
    "        gfp_protein.version = \"1\"\n",
    "        doc.addComponentDefinition(gfp_protein)\n",
    "\n",
    "        gfp_fc = sbol2.FunctionalComponent(gfp_fc_label)\n",
    "        gfp_fc.version = \"1\"\n",
    "        gfp_fc.definition = gfp_protein.identity\n",
    "        gfp_fc.direction = sbol2.SBOL_DIRECTION_OUT\n",
    "        gfp_fc.access = sbol2.SBOL_ACCESS_PUBLIC\n",
    "\n",
    "\n",
    "        doc.add(measurement3)\n",
    "        m = gfp_fc.measurements.add(measurement3)\n",
    "\n",
    "        sample_md.functionalComponents.add(gfp_fc)\n",
    "        # exp_attachment = sbol2.Attachment(exp_label)\n",
    "        # exp_attachment.name = f'fragmentation expression at rLP5 for sample {i}'\n",
    "        # exp_attachment.description = 'CSV including the gene expression of the sequence: RNA_exp1, RNA_exp2, and its average.'\n",
    "        # exp_attachment.source = 'CSV_LINK_HERE' # update when added attachment to SBOL collection\n",
    "        # exp_attachment.format = 'https://identifiers.org/edam/format_3752'\n",
    "        # doc.addAttachment(exp_attachment)\n",
    "\n",
    "\n",
    "        if (i == 0):\n",
    "            break\n",
    "\n",
    "    report = doc.validate()\n",
    "    if (report == 'Valid.'):\n",
    "        doc.write(file_path)\n",
    "    else:\n",
    "        print(report)\n",
    "    return doc, chassis_label_MD\n",
    "    return \"Success\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6bf198d9-d05a-4a6a-9d2e-fad529266188",
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_genome_file_name = \"E. coli.fasta\"\n",
    "env_email = os.getenv(\"SYNBIO_EMAIL\")\n",
    "env_password = os.getenv(\"SYNBIO_PASSWORD\")\n",
    "\n",
    "username = \"cywong\"\n",
    "output_name = 'ecolipromoter_one.xml'\n",
    "id = \"Ecolipromoterexpdataalt\"\n",
    "name = \"E coli promoter data exploration alternative\"\n",
    "description = \"A collection containing the extracted E coli data from paper\"\n",
    "sbol_file_name = output_name\n",
    "imported_sbol_file_name = \"promoters_import.xml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0de59a6b-c443-4788-a8f0-0d1235fab7e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "doc, chassis_label = create_sbol_attempt2_doc(df, os.path.join(sbol_path, output_name))       \n",
    "\n",
    "#create_synbio_collection(env_email, env_password, os.path.join(sbol_path, sbol_file_name), id, name, description)\n",
    "# # # #partshop_attach_exp_data(username, id, attachment_file_names, env_email, env_password, exp_labels)\n",
    "# partshop_attach_genome_to_md(username, id, os.path.join(attachments_path, ecoli_genome_file_name), env_email, env_password, chassis_label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "36c0cfd1-3334-4808-bb9e-2b20ef3c35b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/Ecolipromoterexpdataalt_collection/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_E_coli_chassis/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_0/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_strain_definition_0/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_strain_definition_2/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_strain_definition_4/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_strain_definition_3/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_5/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_4/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_3/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_2/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_strain_definition_1/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_strain_definition_5/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_LB/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_1/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_gfp_3/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_circuit_definition_1/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_gfp_0/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_gfp_5/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_circuit_definition_2/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_gfp_2/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_gfp_1/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_circuit_definition_3/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_gfp_4/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_circuit_definition_5/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_circuit_definition_4/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ComponentDefinition_circuit_definition_0/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/Sequence_promoter_seq_3/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/Sequence_promoter_seq_2/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/Sequence_promoter_seq_4/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/Sequence_promoter_seq_5/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/Sequence_promoter_seq_0/1\n",
      "https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/Sequence_promoter_seq_1/1\n"
     ]
    }
   ],
   "source": [
    "partshop_pull(env_email, env_password, username, id, os.path.join(downloaded_sbol_path, imported_sbol_file_name), version=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a7df4bf5-981f-44ff-8a1a-573dd9996bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# download_all_attachments(env_email, env_password, doc, pulled_attachments_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "48b8bd54-df14-4eba-a637-138bb295fcb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_df_from_sbol_doc(doc, pulled_attachments_path):\n",
    "    df_new = pd.DataFrame()\n",
    "    for exp in doc.experimentalData:\n",
    "        attachment = doc.get(exp.attachments[0])\n",
    "        sample_design = doc.get(exp.wasDerivedFrom[0])\n",
    "    \n",
    "        # may need to be changed\n",
    "        strain = doc.get(sample_design).modules[0] if 'strain' in doc.get(sample_design).modules[0].identity else doc.get(sample_design).modules[1]\n",
    "        promoter =  doc.get(doc.get(strain.definition).functionalComponents[0].definition)\n",
    "        promoter_seq = doc.get(promoter.sequences[0]).elements\n",
    "        \n",
    "        df1 = pd.read_csv(os.path.join(pulled_attachments_path, attachment.name))\n",
    "        df1['Sequence'] = promoter_seq\n",
    "        df_new = pd.concat([df_new, df1], ignore_index=True)\n",
    "\n",
    "    return df_new\n",
    "\n",
    "\n",
    "# for mod in doc.moduleDefinitions:\n",
    "#     print(\"Mod: \", mod.identity)\n",
    "# for attachment in doc.attachments:\n",
    "#     print(\"Attachment: \", attachment.source) \n",
    "    \n",
    "# for seq in doc.sequences:\n",
    "#     print(\"Seq:\", seq.elements) \n",
    "\n",
    "doc = sbol2.Document()\n",
    "doc.read(os.path.join(downloaded_sbol_path, imported_sbol_file_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e292392c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Model(object):\n",
    "    __metaclass__ = ABCMeta\n",
    "\n",
    "    @abstractmethod\n",
    "    def __init__(self, **hyperparameters):\n",
    "        pass\n",
    "\n",
    "    @abstractmethod\n",
    "    def train(self, X, y, validation_data):\n",
    "        pass\n",
    "\n",
    "    @abstractmethod\n",
    "    def predict(self, X):\n",
    "        pass\n",
    "\n",
    "    def test(self, X, y):\n",
    "        return self.evaluate(X, y)\n",
    "        # return ClassificationResult(y, self.predict(X))\n",
    "\n",
    "    def score(self, X, y):\n",
    "        pass\n",
    "\n",
    "class DecisionTree(Model):\n",
    "\n",
    "    def __init__(self):\n",
    "        self.classifier = scikit_DecisionTree()\n",
    "\n",
    "    def train(self, X, y, validation_data=None):\n",
    "        self.classifier.fit(X, y)\n",
    "\n",
    "    def predict(self, X):\n",
    "        predictions = np.asarray(self.classifier.predict_proba(X))[..., 1]\n",
    "        if len(predictions.shape) == 2:  # multitask\n",
    "            predictions = predictions.T\n",
    "        else:  # single-task\n",
    "            predictions = np.expand_dims(predictions, 1)\n",
    "        return predictions\n",
    "    \n",
    "\n",
    "class RandomForestRegression(DecisionTree):\n",
    "    def __init__(self):\n",
    "        self.regressor = RandomForestRegressor(n_estimators=100)\n",
    "\n",
    "    def train(self, X, y, validation_data=None):\n",
    "        # X shape: n_samples, n_features\n",
    "        # y shape: n_samples\n",
    "        self.regressor.fit(X, y)\n",
    "\n",
    "    def predict(self, X):\n",
    "        return self.regressor.predict(X)\n",
    "\n",
    "    def score(self, X, y):\n",
    "        predictions = np.squeeze(self.regressor.predict(X))\n",
    "        return np.corrcoef(predictions, y)[0,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eef3a6bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_hot_encode_modern(sequences):\n",
    "    sequences_array = np.array([[char for char in seq] for seq in sequences])\n",
    "    sequence_length = sequences_array.shape[1]\n",
    "    num_samples = sequences_array.shape[0]\n",
    "\n",
    "    defined_categories = ['A', 'C', 'G', 'T', 'N']\n",
    "    ohe = OneHotEncoder(sparse_output=False,\n",
    "                        categories=[defined_categories] * sequence_length,\n",
    "                        dtype=np.float32) \n",
    "\n",
    "    return ohe.fit_transform(sequences_array)\n",
    "\n",
    "def pad_sequence(seq, max_length):\n",
    "\tif len(seq) > max_length:\n",
    "\t\tdiff = len(seq) - max_length\n",
    "\t\ttrim_length = int(diff / 2)\n",
    "\t\tseq = seq[trim_length : -(trim_length + diff%2)]\n",
    "\telse:\n",
    "\t\tseq = seq.center(max_length, 'N')\n",
    "\treturn seq\n",
    "\n",
    "def process_seqs(df, seq_length, seq_col_name, y_col_name):\n",
    "\tpadded_seqs = [pad_sequence(x, seq_length) for x in df[seq_col_name]]\n",
    "\tX = one_hot_encode_modern(np.array(padded_seqs))\n",
    "\ty = np.array(df[y_col_name], dtype=np.float32)\n",
    "\treturn X, y\n",
    "\n",
    "\n",
    "# def split_data_by_peak(data_df):\n",
    "#     np.random.seed(123)\n",
    "\n",
    "#     # 1. Create 'peak_name' column (if not already present)\n",
    "#     # Assuming peak_start and peak_end are numeric, convert to string for concatenation\n",
    "#     data_df['peak_name'] = data_df['peak_start'].astype(str) + 'to' + data_df['peak_end'].astype(str)\n",
    "\n",
    "#     # 2. Get unique peak names\n",
    "#     peak_names = data_df['peak_name'].unique()\n",
    "\n",
    "#     # 3. Randomly sample train peak names (90%)\n",
    "#     train_size = int(0.90 * len(peak_names))\n",
    "#     train_peak_names = np.random.choice(peak_names, size=train_size, replace=False)\n",
    "\n",
    "#     # 4. Determine test peak names\n",
    "#     test_peak_names = np.array([p for p in peak_names if p not in train_peak_names])\n",
    "\n",
    "\n",
    "#     data_train = data_df[data_df['peak_name'].isin(train_peak_names)].copy()\n",
    "#     data_test = data_df[data_df['peak_name'].isin(test_peak_names)].copy()\n",
    "\n",
    "#     return data_train, data_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8c0490a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df = pd.read_csv(os.path.join(model_data_path, \"tss_expression_model_format_train_genome_split.txt\"), sep='\\t', header=None, names=[\"Sequence\", \"Expression\"])\n",
    "test_df = pd.read_csv(os.path.join(model_data_path, \"tss_expression_model_format_test_genome_split.txt\"), sep='\\t',  header=None, names=[\"Sequence\", \"Expression\"])\n",
    "\n",
    "X_train, y_train = process_seqs(train_df, 150, \"Sequence\", \"Expression\")\n",
    "X_test, y_test = process_seqs(test_df, 150, \"Sequence\", \"Expression\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "9826fbb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_RandomForest(X_train, X_test, y_train, y_test):\n",
    "    print(\"Running random forest regression...\")\n",
    "    model = RandomForestRegression()\n",
    "    model.train(X_train, y_train)\n",
    "    predictions = model.predict(X_test)\n",
    "    \n",
    "    with open(\"outputs.txt\", 'w') as outfile:\n",
    "        for i in range(len(predictions)):\n",
    "            outfile.write(str(float(predictions[i])) + '\\t' +\n",
    "                      str(float(y_test[i])) + '\\n')\n",
    "    \n",
    "    score = model.score(X_test, y_test)\n",
    "    print(\"Score:\", score)\n",
    "    return model, predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "cb56143f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running random forest regression...\n",
      "Score: 0.5577909140696771\n"
     ]
    }
   ],
   "source": [
    "model, predictions = train_RandomForest(X_train, X_test, y_train, y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5ab152ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean Squared Error (MSE): 17.536704867201603\n",
      "Root Mean Square Error (RMSE): 4.187684905434219\n",
      "R^2: 0.26785076457523127\n"
     ]
    }
   ],
   "source": [
    "mse = mean_squared_error(y_test, predictions)\n",
    "r_squared = r2_score(y_test, predictions)\n",
    "rmse = np.sqrt(mse)\n",
    "\n",
    "print(f\"Mean Squared Error (MSE): {mse}\")\n",
    "print(f\"Root Mean Square Error (RMSE): {rmse}\")\n",
    "print(f\"R^2: {r_squared}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "11b575e9-ebaf-4789-9bc9-9b67f80b4659",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.iloc[0:5000].sample(frac=1, random_state=42).reset_index(drop=True)\n",
    "df['strand'] = df['strand'].map({'+': 0, '-': 1})\n",
    "df.drop(['start', 'end'], axis=1, inplace=True)\n",
    "\n",
    "\n",
    "columns_to_scale = ['RNA_exp_1', 'DNA_sum_1', 'DNA_sum_2', 'DNA_ave', 'variation']\n",
    "scaler = StandardScaler()\n",
    "df[columns_to_scale] = scaler.fit_transform(df[columns_to_scale])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "143f885b-13a5-4b8c-a263-f086a36bd56e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a94af75e-ea56-458b-a2ad-703fb4e1d37e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fragment</th>\n",
       "      <th>RNA_exp_1</th>\n",
       "      <th>RNA_exp_2</th>\n",
       "      <th>RNA_exp_ave</th>\n",
       "      <th>DNA_sum_1</th>\n",
       "      <th>DNA_sum_2</th>\n",
       "      <th>DNA_ave</th>\n",
       "      <th>num_mapped_barcodes</th>\n",
       "      <th>num_integrated_barcodes</th>\n",
       "      <th>strand</th>\n",
       "      <th>variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CGTTTCGGGGTTAAATCCGCAAACAGGATAAACATGCCAGTCACTA...</td>\n",
       "      <td>-0.288873</td>\n",
       "      <td>1.377625</td>\n",
       "      <td>0.865143</td>\n",
       "      <td>-0.850522</td>\n",
       "      <td>-0.850522</td>\n",
       "      <td>-0.850522</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.264872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TATTCCCTTCTACAATTTCAGCGCCCGCTTGCACAAACGCTTTATC...</td>\n",
       "      <td>-0.175605</td>\n",
       "      <td>0.768593</td>\n",
       "      <td>0.699418</td>\n",
       "      <td>1.564873</td>\n",
       "      <td>1.564873</td>\n",
       "      <td>1.564873</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.358652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATTTCATGATCCTGCGTCCACAGCAGAAGCGCACCAAAGAACACAA...</td>\n",
       "      <td>-0.086170</td>\n",
       "      <td>0.850130</td>\n",
       "      <td>0.849773</td>\n",
       "      <td>0.699372</td>\n",
       "      <td>0.699372</td>\n",
       "      <td>0.699372</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.143496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TATTGATTCGGCGGATGGTTTGCCGATGGTGGTGTACAACATTCCA...</td>\n",
       "      <td>-0.186059</td>\n",
       "      <td>0.936897</td>\n",
       "      <td>0.770759</td>\n",
       "      <td>-0.787677</td>\n",
       "      <td>-0.787677</td>\n",
       "      <td>-0.787677</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.592592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCCATCCCGACCGAGGCACTACGCCAGTTGATGTCGTGGGATTGGC...</td>\n",
       "      <td>-0.088598</td>\n",
       "      <td>0.502205</td>\n",
       "      <td>0.672836</td>\n",
       "      <td>-0.810974</td>\n",
       "      <td>-0.810974</td>\n",
       "      <td>-0.810974</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.912471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>CCGATTTCGCGGATCGCCAGGCGGTATCCGTTCAACGGAGAGATGG...</td>\n",
       "      <td>-0.217364</td>\n",
       "      <td>0.518143</td>\n",
       "      <td>0.523023</td>\n",
       "      <td>0.136090</td>\n",
       "      <td>0.136090</td>\n",
       "      <td>0.136090</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.072712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>AGGGGCGACGGTTTTTGGCTATCAGGTGATGGACCCGGAACGCTTT...</td>\n",
       "      <td>0.122469</td>\n",
       "      <td>0.556262</td>\n",
       "      <td>0.958491</td>\n",
       "      <td>-0.717651</td>\n",
       "      <td>-0.717651</td>\n",
       "      <td>-0.717651</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.405852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>CGGACTAAATACTGATGAGTTAAACTCCATCGTCCATTCGGTCCAG...</td>\n",
       "      <td>-0.195258</td>\n",
       "      <td>0.834329</td>\n",
       "      <td>0.708204</td>\n",
       "      <td>-0.768300</td>\n",
       "      <td>-0.768300</td>\n",
       "      <td>-0.768300</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.283023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>CGTGGTTTGCGTGGATCTTCGGTGCGTTGTGCTGGATGACGACCTT...</td>\n",
       "      <td>-0.071135</td>\n",
       "      <td>0.468962</td>\n",
       "      <td>0.677612</td>\n",
       "      <td>-0.268633</td>\n",
       "      <td>-0.268633</td>\n",
       "      <td>-0.268633</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.381039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>TGGGCTTAAGGCGGAAGTTTTCCCGAATGCGGATCTCCGGCAGCAG...</td>\n",
       "      <td>-0.199280</td>\n",
       "      <td>0.599828</td>\n",
       "      <td>0.586025</td>\n",
       "      <td>0.883662</td>\n",
       "      <td>0.883662</td>\n",
       "      <td>0.883662</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.959707</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               fragment  RNA_exp_1  RNA_exp_2  \\\n",
       "0     CGTTTCGGGGTTAAATCCGCAAACAGGATAAACATGCCAGTCACTA...  -0.288873   1.377625   \n",
       "1     TATTCCCTTCTACAATTTCAGCGCCCGCTTGCACAAACGCTTTATC...  -0.175605   0.768593   \n",
       "2     ATTTCATGATCCTGCGTCCACAGCAGAAGCGCACCAAAGAACACAA...  -0.086170   0.850130   \n",
       "3     TATTGATTCGGCGGATGGTTTGCCGATGGTGGTGTACAACATTCCA...  -0.186059   0.936897   \n",
       "4     GCCATCCCGACCGAGGCACTACGCCAGTTGATGTCGTGGGATTGGC...  -0.088598   0.502205   \n",
       "...                                                 ...        ...        ...   \n",
       "4995  CCGATTTCGCGGATCGCCAGGCGGTATCCGTTCAACGGAGAGATGG...  -0.217364   0.518143   \n",
       "4996  AGGGGCGACGGTTTTTGGCTATCAGGTGATGGACCCGGAACGCTTT...   0.122469   0.556262   \n",
       "4997  CGGACTAAATACTGATGAGTTAAACTCCATCGTCCATTCGGTCCAG...  -0.195258   0.834329   \n",
       "4998  CGTGGTTTGCGTGGATCTTCGGTGCGTTGTGCTGGATGACGACCTT...  -0.071135   0.468962   \n",
       "4999  TGGGCTTAAGGCGGAAGTTTTCCCGAATGCGGATCTCCGGCAGCAG...  -0.199280   0.599828   \n",
       "\n",
       "      RNA_exp_ave  DNA_sum_1  DNA_sum_2   DNA_ave  num_mapped_barcodes  \\\n",
       "0        0.865143  -0.850522  -0.850522 -0.850522                    1   \n",
       "1        0.699418   1.564873   1.564873  1.564873                    1   \n",
       "2        0.849773   0.699372   0.699372  0.699372                    2   \n",
       "3        0.770759  -0.787677  -0.787677 -0.787677                    3   \n",
       "4        0.672836  -0.810974  -0.810974 -0.810974                    1   \n",
       "...           ...        ...        ...       ...                  ...   \n",
       "4995     0.523023   0.136090   0.136090  0.136090                    1   \n",
       "4996     0.958491  -0.717651  -0.717651 -0.717651                    1   \n",
       "4997     0.708204  -0.768300  -0.768300 -0.768300                    1   \n",
       "4998     0.677612  -0.268633  -0.268633 -0.268633                    1   \n",
       "4999     0.586025   0.883662   0.883662  0.883662                    1   \n",
       "\n",
       "      num_integrated_barcodes  strand  variation  \n",
       "0                           1     NaN   4.264872  \n",
       "1                           1     NaN  -0.358652  \n",
       "2                           2     NaN  -1.143496  \n",
       "3                           1     NaN   0.592592  \n",
       "4                           1     NaN   0.912471  \n",
       "...                       ...     ...        ...  \n",
       "4995                        1     NaN  -1.072712  \n",
       "4996                        1     NaN   2.405852  \n",
       "4997                        1     NaN   0.283023  \n",
       "4998                        1     NaN   1.381039  \n",
       "4999                        1     NaN  -0.959707  \n",
       "\n",
       "[5000 rows x 11 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6368cbb7-615d-4ae1-831d-e00891f1b7bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = []\n",
    "for i in range(len(df)):\n",
    "    seq = pad_sequence(df.loc[i, 'fragment'], 200)\n",
    "    ohe = one_hot_encode_modern(seq)\n",
    "    a.append(ohe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6d328caa-daac-450d-91d9-1ef63b3f3867",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = np.array(a)\n",
    "X1 = X.reshape(X.shape[0], -1)\n",
    "X2 = df.drop(['RNA_exp_ave', 'fragment'], axis=1).values\n",
    "X = np.concatenate([X1, X2], axis=1)\n",
    "y = np.log1p(df['RNA_exp_ave'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "738c09ab-7b61-4b6e-99e8-81b41c9c9b6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    5000.000000\n",
       "mean        1.073838\n",
       "std         2.296893\n",
       "min         0.109815\n",
       "25%         0.743188\n",
       "50%         0.876965\n",
       "75%         1.039300\n",
       "max        98.037450\n",
       "Name: RNA_exp_ave, dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " df['RNA_exp_ave'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "88a9f35d-a919-4fd8-b1ff-8e169f10ec69",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "xgb_model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "25c4f162-f50b-4994-b0a9-abacb4170c79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.0, 0.0, 1.0, ..., 1, 1, -0.464931406881896],\n",
       "       [1.0, 0.0, 0.0, ..., 1, 0, 0.47187132891183875],\n",
       "       [0.0, 1.0, 0.0, ..., 1, 0, -0.45230363693030246],\n",
       "       ...,\n",
       "       [1.0, 0.0, 0.0, ..., 1, 0, 0.2830228179142664],\n",
       "       [0.0, 0.0, 1.0, ..., 1, 1, 1.3810387050466513],\n",
       "       [0.0, 0.0, 1.0, ..., 1, 0, -0.9597073076872386]], dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "226ea739-e41a-46b1-98b8-cad2a3c3088d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 3 folds for each of 16 candidates, totalling 48 fits\n"
     ]
    }
   ],
   "source": [
    "param_grid = {\n",
    "    'n_estimators': [50, 100],\n",
    "    'max_depth': [3, 7],\n",
    "    'learning_rate': [0.01, 0.1],\n",
    "    'colsample_bytree': [0.8, 1.0]\n",
    "}\n",
    "\n",
    "grid_search = GridSearchCV(\n",
    "    estimator=xgb_model,\n",
    "    param_grid=param_grid,\n",
    "    scoring='neg_mean_squared_error',  \n",
    "    cv=3,\n",
    "    verbose=1,\n",
    ")\n",
    "\n",
    "grid = grid_search.fit(X_train, y_train)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b1f48c49-1a43-41d1-879d-c381cf4ee734",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best parameters: {'colsample_bytree': 0.8, 'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100}\n",
      "Test MSE: 0.005051369635567891\n",
      "R sq: 0.935688593845861\n"
     ]
    }
   ],
   "source": [
    "print(\"Best parameters:\", grid_search.best_params_)\n",
    "\n",
    "best_model = grid_search.best_estimator_\n",
    "y_pred = best_model.predict(X_test)\n",
    "\n",
    "mse = mean_squared_error(y_test, y_pred)\n",
    "r_sq = r2_score(y_test, y_pred)\n",
    "print(\"Test MSE:\", mse)\n",
    "print(\"R sq:\", r_sq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "002109e6-091b-41ee-ac79-a454a9d004f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "xgb_model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42, colsample_bytree=1, learning_rate=0.1, max_depth=3, n_estimators=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d5d6c413-fa53-4b38-8139-8f700e9a36b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "xgb_model.fit(X_train, y_train)\n",
    "\n",
    "preds = xgb_model.predict(X_test)\n",
    "\n",
    "r_sq = r2_score(y_test, preds) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "59dad312-a450-4c42-abcb-20f5887139fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-3.122416624183555"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r_sq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f17db0ed-259e-4f14-bbc8-15d7c79d8728",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
