{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6169c8c4-6c46-41c1-9c84-3bff9d2e2c8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sbol2\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "778cff54-44b6-48a4-871e-efc95f7c6631",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[2], line 7\u001b[0m\n\u001b[0;32m      3\u001b[0m sbol2\u001b[38;5;241m.\u001b[39msetHomespace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttp://github.com/cywlol/promoters\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m      4\u001b[0m version \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m1.0\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m----> 7\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, row \u001b[38;5;129;01min\u001b[39;00m df\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[0;32m      8\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m (i \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m10\u001b[39m):\n\u001b[0;32m      9\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "def loop_over_df(df):\n",
    "    doc = sbol2.Document()\n",
    "    sbol2.setHomespace('http://github.com/cywlol/promoters')\n",
    "    version = '1.0'\n",
    "    \n",
    "    for i, row in df.iterrows():\n",
    "        if (i == 10):\n",
    "            break\n",
    "            \n",
    "        fragment_seq = row[\"fragment\"]\n",
    "        rna_exp_1 = row[\"RNA_exp_1\"]\n",
    "        rna_exp_2 = row[\"RNA_exp_2\"]\n",
    "        rna_exp_ave = row[\"RNA_exp_ave\"]\n",
    "        dna_sum_1 = row[\"DNA_sum_1\"]\n",
    "        dna_sum_2 = row[\"DNA_sum_2\"]\n",
    "        dna_ave = row[\"DNA_ave\"]\n",
    "        num_integrated = row[\"num_integrated_barcodes\"]\n",
    "        start = row[\"start\"]\n",
    "        end = row[\"end\"]\n",
    "        strand = row[\"strand\"]\n",
    "        variation = row[\"variation\"]\n",
    "        media_label = \"LB\"\n",
    "    \n",
    "        media_label_MD = f\"media_{i}\"\n",
    "        chassis_label_MD = f\"chassis_{i}\"\n",
    "        promoter_label_CD = f\"promoter_{i}\"\n",
    "        promoter_seq = f\"promoter_seq_{i}\"\n",
    "        engr_region_label_CD = f\"engr_region_{i}\"\n",
    "        strain_label_CD = f\"strain{i}\"\n",
    "        sample_design_label_CD = f\"sample_design_{i}\"\n",
    "        location_promoter_annotation = f\"location_promoter_annotation_{i}\"\n",
    "        location_promoter_label = f\"location_promoter_label_{i}\"\n",
    "        attachment_label = f\"ecoli_mg1655_fasta_{i}\"\n",
    "        data_label = f\"exp_data_{i}\"\n",
    "        \n",
    "        # Media (MD)\n",
    "        media_md = sbol2.ModuleDefinition(media_label_MD)\n",
    "        doc.addComponentDefinition(media_md)\n",
    "        media_cd.addRole(\"http://identifiers.org/NCIT/NCIT:C85504\") \n",
    "        \n",
    "        # Chassis \n",
    "        chassis_cd = sbol2.ModuleDefinition(chassis_label_MD) \n",
    "        doc.addComponentDefinition(chassis_cd)\n",
    "    \n",
    "        fasta_attachment = sbol2.Attachment(attachment_label)\n",
    "        fasta_attachment.name = 'E. coli MG1655 genome'\n",
    "        fasta_attachment.description = 'E. coli MG1655 genome sequence'\n",
    "        fasta_attachment.source = 'https:/0www.ncbi.nlm.nih.gov/nuccore/U00096.3?report=fasta'\n",
    "        fasta_attachment.format = 'https://identifiers.org/edam/format_1929'\n",
    "        chassis_cd.attachments = [fasta_attachment.persistentIdentity]\n",
    "        doc.addAttachment(fasta_attachment)\n",
    "        \n",
    "        # Promoter and Sequence \n",
    "        promoter_cd = sbol2.ComponentDefinition(promoter_label_CD, sbol2.BIOPAX_DNA)\n",
    "        promoter_cd.roles = [sbol2.SO_PROMOTER]\n",
    "        seq = sbol2.Sequence(promoter_seq, fragment_seq, sbol2.SBOL_ENCODING_IUPAC)\n",
    "        doc.addSequence(seq)\n",
    "        promoter_cd.sequences = [seq.persistentIdentity]\n",
    "        doc.addComponentDefinition(promoter_cd)\n",
    "    \n",
    "        annotation = sbol2.SequenceAnnotation(\"promoter_location\")\n",
    "        range = sbol2.Range(\"prange\", start, end)\n",
    "    \n",
    "        if (strand == \"-\"):\n",
    "            range.orientation = sbol2.SBOL_ORIENTATION_REVERSE_COMPLEMENT\n",
    "                \n",
    "        annotation.locations.add(range)\n",
    "        promoter_cd.sequenceAnnotations.add(annotation)\n",
    "        \n",
    "        # Engineered Region \n",
    "        engineered_cd = sbol2.ComponentDefinition(engr_region_label_CD, sbol2.BIOPAX_DNA)\n",
    "        engineered_cd.roles = [\"https://identifiers.org/so/SO:0000804\"]\n",
    "        sub = engineered_cd.components.create('promoter')\n",
    "        sub.definition = promoter_cd.persistentIdentity\n",
    "        doc.addComponentDefinition(engineered_cd)\n",
    "        # No sequence?\n",
    "        \n",
    "        # Strain \n",
    "        \n",
    "        strain_cd = sbol2.ModuleDefinition(strain_label_CD)\n",
    "        doc.addComponentDefinition(strain_cd)\n",
    "        strain_c1 = strain_cd.modules.create('chassis')\n",
    "        strain_c1.definition = chassis_cd.persistentIdentity\n",
    "        \n",
    "        strain_c2 = strain_cd.functionalComponents.create('engineered_region')\n",
    "        strain_c2.definition = engineered_cd.persistentIdentity\n",
    "        \n",
    "        # Sample Design  \n",
    "        sample_md = sbol2.ModuleDefinition(sample_design_label_CD)\n",
    "        doc.addModuleDefinition(sample_md)\n",
    "        \n",
    "        fc_strain = sample_md.functionalComponents.create('strain')\n",
    "        fc_strain.definition = strain_cd.persistentIdentity\n",
    "        \n",
    "        fc_media = sample_md.modules.create('media')\n",
    "        fc_media.definition = media_cd.persistentIdentity\n",
    "    \n",
    "        # Experiment and Measurement Data\n",
    "        # exp = sbol2.ExperimentalData('data')\n",
    "        # exp.wasDerivedFrom.append(sample_md.persistentIdentity)\n",
    "        # measure = sbol2.Measurement('rna_exp_1', rna_exp_1)\n",
    "        # exp.addPropertyValue('rna_exp_1',measure)\n",
    "        # doc.validate()\n",
    "    \n",
    "    doc.write('test.xml')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "db274a9c-655b-45a9-bdfb-8e062a5a023c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "98c2233f-51ee-4164-aa8e-205713ed7c0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "loop_over_df(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "346a1555-2376-4829-8045-c25886afb586",
   "metadata": {},
   "outputs": [],
   "source": [
    "def loop_over_df(data, column_names):\n",
    "import pandas as pd\n",
    "import sbol2\n",
    "\n",
    "# Sample dataframe for demonstration\n",
    "data = {\n",
    "    \"RNA_exp_1\": [1.316755] * 15,\n",
    "    \"RNA_exp_2\": [1.635498] * 15,\n",
    "    \"RNA_exp_ave\": [1.476127] * 15,\n",
    "    \"DNA_sum_1\": [1.006226] * 15,\n",
    "    \"DNA_sum_2\": [1.006226] * 15,\n",
    "    \"DNA_ave\": [1.006226] * 15,\n",
    "    \"Start\": [1838405 + i for i in range(15)],\n",
    "    \"End\": [1838702 + i for i in range(15)],\n",
    "    \"Strand\": [\"-\"] * 15,\n",
    "    \"Variation\": [0.312743] * 15,\n",
    "    \"Mapped\": [1] * 15,\n",
    "    \"Integrated\": [1] * 15\n",
    "}\n",
    "df = pd.DataFrame(data)\n",
    "\n",
    "# SBOL setup\n",
    "sbol2.setHomespace('http://github.com/cywlol/promoters')\n",
    "doc = sbol2.Document()\n",
    "version = '1.0'\n",
    "media_label = \"LB\"\n",
    "\n",
    "# Media (MD)\n",
    "media_cd = sbol2.ModuleDefinition(media_label, version)\n",
    "doc.addModuleDefinition(media_cd)\n",
    "\n",
    "# Chassis\n",
    "chassis_cd = sbol2.ModuleDefinition('Ecoli_K12', version)\n",
    "doc.addModuleDefinition(chassis_cd)\n",
    "\n",
    "def loop_over_df(df):\n",
    "        \n",
    "    doc = sbol2.Document()\n",
    "    sbol2.setHomespace('http://github.com/cywlol/promoters')\n",
    "    version = '1.0'\n",
    "\n",
    "\n",
    "for i, row in df.iterrows():\n",
    "    if (i == 10):\n",
    "        break\n",
    "        \n",
    "    fragment_seq = row[\"fragment\"]\n",
    "    rna_exp_1 = row[\"RNA_exp_1\"]\n",
    "    rna_exp_2 = row[\"RNA_exp_2\"]\n",
    "    rna_exp_ave = row[\"RNA_exp_ave\"]\n",
    "    dna_sum_1 = row[\"DNA_sum_1\"]\n",
    "    dna_sum_2 = row[\"DNA_sum_2\"]\n",
    "    dna_ave = row[\"DNA_ave\"]\n",
    "    num_integrated = row[\"num_integrated_barcodes\"]\n",
    "    start = row[\"start\"]\n",
    "    end = row[\"end\"]\n",
    "    strand = row[\"strand\"]\n",
    "    variation = row[\"variation\"]\n",
    "    media_label = \"LB\"\n",
    "\n",
    "    media_label_MD = f\"media_{i}\"\n",
    "    chassis_label_MD = f\"chassis_{i}\"\n",
    "    promoter_label_CD = f\"promoter_{i}\"\n",
    "    promoter_seq = f\"promoter_seq_{i}\"\n",
    "    engr_region_label_CD = f\"engr_region_{i}\"\n",
    "    strain_label_CD = f\"strain{i}\"\n",
    "    sample_design_label_CD = f\"sample_design_{i}\"\n",
    "    location_promoter_annotation = f\"location_promoter_annotation_{i}\"\n",
    "    location_promoter_label = f\"location_promoter_label_{i}\"\n",
    "    attachment_label = f\"ecoli_mg1655_fasta_{i}\"\n",
    "    data_label = f\"exp_data_{i}\"\n",
    "    \n",
    "    # Media (MD)\n",
    "    media_md = sbol2.ModuleDefinition(media_label_MD)\n",
    "    doc.addComponentDefinition(media_md)\n",
    "    media_cd.addRole(\"http://identifiers.org/NCIT/NCIT:C85504\") \n",
    "    \n",
    "    # Chassis \n",
    "    chassis_cd = sbol2.ModuleDefinition(chassis_label_MD) \n",
    "    doc.addComponentDefinition(chassis_cd)\n",
    "\n",
    "    fasta_attachment = sbol2.Attachment(attachment_label)\n",
    "    fasta_attachment.name = 'E. coli MG1655 genome'\n",
    "    fasta_attachment.description = 'E. coli MG1655 genome sequence'\n",
    "    fasta_attachment.source = 'https:/0www.ncbi.nlm.nih.gov/nuccore/U00096.3?report=fasta'\n",
    "    fasta_attachment.format = 'https://identifiers.org/edam/format_1929'\n",
    "    chassis_cd.attachments = [fasta_attachment.persistentIdentity]\n",
    "    doc.addAttachment(fasta_attachment)\n",
    "    \n",
    "    # Promoter and Sequence \n",
    "    promoter_cd = sbol2.ComponentDefinition(promoter_label_CD, sbol2.BIOPAX_DNA)\n",
    "    promoter_cd.roles = [sbol2.SO_PROMOTER]\n",
    "    seq = sbol2.Sequence(promoter_seq, fragment_seq, sbol2.SBOL_ENCODING_IUPAC)\n",
    "    doc.addSequence(seq)\n",
    "    promoter_cd.sequences = [seq.persistentIdentity]\n",
    "    doc.addComponentDefinition(promoter_cd)\n",
    "\n",
    "    annotation = sbol2.SequenceAnnotation(\"promoter_location\")\n",
    "    range = sbol2.Range(\"prange\", start, end)\n",
    "\n",
    "    if (strand == \"-\"):\n",
    "        range.orientation = sbol2.SBOL_ORIENTATION_REVERSE_COMPLEMENT\n",
    "            \n",
    "    annotation.locations.add(range)\n",
    "    promoter_cd.sequenceAnnotations.add(annotation)\n",
    "    \n",
    "    # Engineered Region \n",
    "    engineered_cd = sbol2.ComponentDefinition(engr_region_label_CD, sbol2.BIOPAX_DNA)\n",
    "    engineered_cd.roles = [\"https://identifiers.org/so/SO:0000804\"]\n",
    "    sub = engineered_cd.components.create('promoter')\n",
    "    sub.definition = promoter_cd.persistentIdentity\n",
    "    doc.addComponentDefinition(engineered_cd)\n",
    "    # No sequence?\n",
    "    \n",
    "    # Strain \n",
    "    \n",
    "    strain_cd = sbol2.ModuleDefinition(strain_label_CD)\n",
    "    doc.addComponentDefinition(strain_cd)\n",
    "    strain_c1 = strain_cd.modules.create('chassis')\n",
    "    strain_c1.definition = chassis_cd.persistentIdentity\n",
    "    \n",
    "    strain_c2 = strain_cd.functionalComponents.create('engineered_region')\n",
    "    strain_c2.definition = engineered_cd.persistentIdentity\n",
    "    \n",
    "    # Sample Design  \n",
    "    sample_md = sbol2.ModuleDefinition(sample_design_label_CD)\n",
    "    doc.addModuleDefinition(sample_md)\n",
    "    \n",
    "    fc_strain = sample_md.functionalComponents.create('strain')\n",
    "    fc_strain.definition = strain_cd.persistentIdentity\n",
    "    \n",
    "    fc_media = sample_md.modules.create('media')\n",
    "    fc_media.definition = media_cd.persistentIdentity\n",
    "\n",
    "    # Experiment and Measurement Data\n",
    "    # exp = sbol2.ExperimentalData('data')\n",
    "    # exp.wasDerivedFrom.append(sample_md.persistentIdentity)\n",
    "    # measure = sbol2.Measurement('rna_exp_1', rna_exp_1)\n",
    "    # exp.addPropertyValue('rna_exp_1',measure)\n",
    "    # doc.validate()\n",
    "\n",
    "\n",
    "loop_over_df(df)\n",
    "doc.validate()\n",
    "doc.write('fragmented_promoters.xml')\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "312038f5-92a7-4a18-afd9-820d44728fb4",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df4' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m loop_over_df(df4, [])\n",
      "\u001b[1;31mNameError\u001b[0m: name 'df4' is not defined"
     ]
    }
   ],
   "source": [
    "loop_over_df(df4, [])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "05b881c7-b711-41ab-af99-5c3e0dcb7245",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['fragment', 'RNA_exp_1', 'RNA_exp_2', 'RNA_exp_ave', 'DNA_sum_1',\n",
       "       'DNA_sum_2', 'DNA_ave', 'num_mapped_barcodes',\n",
       "       'num_integrated_barcodes', 'start', 'end', 'strand', 'variation'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"frag-rLP5_LB_expression.txt\", delimiter=\" \")\n",
    "df.columns\n",
    "# Expression levels from genomic fragment MPRA (random 200–300 bp sheared fragments) in LB media\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87ab79e1-ab65-43f4-9278-ad6fb3b00361",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5ce856ae-3aa8-4f03-a52b-9a8918d8e291",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fragment</th>\n",
       "      <th>RNA_exp_1</th>\n",
       "      <th>RNA_exp_2</th>\n",
       "      <th>RNA_exp_ave</th>\n",
       "      <th>DNA_sum_1</th>\n",
       "      <th>DNA_sum_2</th>\n",
       "      <th>DNA_ave</th>\n",
       "      <th>num_mapped_barcodes</th>\n",
       "      <th>num_integrated_barcodes</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ATCTTTAAATATTTAAAAGAACTGGGAGTACCCGCGAGTGCCGCTG...</td>\n",
       "      <td>0.890051</td>\n",
       "      <td>0.922716</td>\n",
       "      <td>0.906383</td>\n",
       "      <td>1.233452</td>\n",
       "      <td>1.233452</td>\n",
       "      <td>1.233452</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1892252</td>\n",
       "      <td>1892540</td>\n",
       "      <td>+</td>\n",
       "      <td>0.051998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGCCGGGATCTCAACTGGCACACTGACAGCAGGCTCAGCAACAGCT...</td>\n",
       "      <td>0.417847</td>\n",
       "      <td>0.500726</td>\n",
       "      <td>0.459286</td>\n",
       "      <td>1.987298</td>\n",
       "      <td>1.987298</td>\n",
       "      <td>1.987298</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3110405</td>\n",
       "      <td>3110614</td>\n",
       "      <td>+</td>\n",
       "      <td>0.261046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTTCCAACCCATGGCGCGTGCGTACATAAAGGTTTCGGACGCGCGC...</td>\n",
       "      <td>1.652097</td>\n",
       "      <td>0.447871</td>\n",
       "      <td>1.049984</td>\n",
       "      <td>1.136735</td>\n",
       "      <td>1.136735</td>\n",
       "      <td>1.136735</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1838405</td>\n",
       "      <td>1838702</td>\n",
       "      <td>-</td>\n",
       "      <td>1.883143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TGACTTCTTCGTGAACTTGCTGGATATGCGTTACGAGTGGAAAGCG...</td>\n",
       "      <td>0.667351</td>\n",
       "      <td>0.607438</td>\n",
       "      <td>0.637395</td>\n",
       "      <td>1.979343</td>\n",
       "      <td>1.979343</td>\n",
       "      <td>1.979343</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4133776</td>\n",
       "      <td>4134009</td>\n",
       "      <td>+</td>\n",
       "      <td>0.135708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TGGGGGGATGTATGGGTACGTTGTAATTAGGGATTTAACGAATTAG...</td>\n",
       "      <td>0.573034</td>\n",
       "      <td>0.632320</td>\n",
       "      <td>0.602677</td>\n",
       "      <td>1.342102</td>\n",
       "      <td>1.342102</td>\n",
       "      <td>1.342102</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2862818</td>\n",
       "      <td>2863076</td>\n",
       "      <td>-</td>\n",
       "      <td>0.142033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318452</th>\n",
       "      <td>ATGCCAGCAGGTCAGTGACCTGGCTGTGAATCGAGATAAAGCCGAT...</td>\n",
       "      <td>0.856074</td>\n",
       "      <td>0.842344</td>\n",
       "      <td>0.849209</td>\n",
       "      <td>6.054634</td>\n",
       "      <td>6.054634</td>\n",
       "      <td>6.054634</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>887886</td>\n",
       "      <td>888093</td>\n",
       "      <td>+</td>\n",
       "      <td>0.023327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318453</th>\n",
       "      <td>TGCCCGTTGATTTTCAGAGAAGGGGAATTAGTACAGCAGACGGGCG...</td>\n",
       "      <td>0.354607</td>\n",
       "      <td>0.589399</td>\n",
       "      <td>0.472003</td>\n",
       "      <td>2.168799</td>\n",
       "      <td>2.168799</td>\n",
       "      <td>2.168799</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3055172</td>\n",
       "      <td>3055425</td>\n",
       "      <td>+</td>\n",
       "      <td>0.733025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318454</th>\n",
       "      <td>GGCCTGGTGGCCTGCATCGCGCTGGCGCTGATTATCGCCACGCTCG...</td>\n",
       "      <td>1.053131</td>\n",
       "      <td>1.695967</td>\n",
       "      <td>1.374549</td>\n",
       "      <td>10.049118</td>\n",
       "      <td>10.049118</td>\n",
       "      <td>10.049118</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3844004</td>\n",
       "      <td>3844212</td>\n",
       "      <td>-</td>\n",
       "      <td>0.687423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318455</th>\n",
       "      <td>AAAATGAAATTGGGCAGTTGAAACCAGACGTTTCGCCCCTATTACA...</td>\n",
       "      <td>0.625138</td>\n",
       "      <td>1.328967</td>\n",
       "      <td>0.977053</td>\n",
       "      <td>3.753320</td>\n",
       "      <td>3.753320</td>\n",
       "      <td>3.753320</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1862773</td>\n",
       "      <td>1863068</td>\n",
       "      <td>-</td>\n",
       "      <td>1.088059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318456</th>\n",
       "      <td>CAATATCGCCAACTATTTTGGTCAGTGGCCGGTGGATTATAAGAGC...</td>\n",
       "      <td>0.889288</td>\n",
       "      <td>0.665998</td>\n",
       "      <td>0.777643</td>\n",
       "      <td>9.331071</td>\n",
       "      <td>9.331071</td>\n",
       "      <td>9.331071</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2900333</td>\n",
       "      <td>2900531</td>\n",
       "      <td>+</td>\n",
       "      <td>0.417133</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>318457 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 fragment  RNA_exp_1  \\\n",
       "0       ATCTTTAAATATTTAAAAGAACTGGGAGTACCCGCGAGTGCCGCTG...   0.890051   \n",
       "1       AGCCGGGATCTCAACTGGCACACTGACAGCAGGCTCAGCAACAGCT...   0.417847   \n",
       "2       CTTCCAACCCATGGCGCGTGCGTACATAAAGGTTTCGGACGCGCGC...   1.652097   \n",
       "3       TGACTTCTTCGTGAACTTGCTGGATATGCGTTACGAGTGGAAAGCG...   0.667351   \n",
       "4       TGGGGGGATGTATGGGTACGTTGTAATTAGGGATTTAACGAATTAG...   0.573034   \n",
       "...                                                   ...        ...   \n",
       "318452  ATGCCAGCAGGTCAGTGACCTGGCTGTGAATCGAGATAAAGCCGAT...   0.856074   \n",
       "318453  TGCCCGTTGATTTTCAGAGAAGGGGAATTAGTACAGCAGACGGGCG...   0.354607   \n",
       "318454  GGCCTGGTGGCCTGCATCGCGCTGGCGCTGATTATCGCCACGCTCG...   1.053131   \n",
       "318455  AAAATGAAATTGGGCAGTTGAAACCAGACGTTTCGCCCCTATTACA...   0.625138   \n",
       "318456  CAATATCGCCAACTATTTTGGTCAGTGGCCGGTGGATTATAAGAGC...   0.889288   \n",
       "\n",
       "        RNA_exp_2  RNA_exp_ave  DNA_sum_1  DNA_sum_2    DNA_ave  \\\n",
       "0        0.922716     0.906383   1.233452   1.233452   1.233452   \n",
       "1        0.500726     0.459286   1.987298   1.987298   1.987298   \n",
       "2        0.447871     1.049984   1.136735   1.136735   1.136735   \n",
       "3        0.607438     0.637395   1.979343   1.979343   1.979343   \n",
       "4        0.632320     0.602677   1.342102   1.342102   1.342102   \n",
       "...           ...          ...        ...        ...        ...   \n",
       "318452   0.842344     0.849209   6.054634   6.054634   6.054634   \n",
       "318453   0.589399     0.472003   2.168799   2.168799   2.168799   \n",
       "318454   1.695967     1.374549  10.049118  10.049118  10.049118   \n",
       "318455   1.328967     0.977053   3.753320   3.753320   3.753320   \n",
       "318456   0.665998     0.777643   9.331071   9.331071   9.331071   \n",
       "\n",
       "        num_mapped_barcodes  num_integrated_barcodes    start      end strand  \\\n",
       "0                         3                        1  1892252  1892540      +   \n",
       "1                         1                        1  3110405  3110614      +   \n",
       "2                         1                        1  1838405  1838702      -   \n",
       "3                         4                        1  4133776  4134009      +   \n",
       "4                         3                        1  2862818  2863076      -   \n",
       "...                     ...                      ...      ...      ...    ...   \n",
       "318452                    3                        1   887886   888093      +   \n",
       "318453                    1                        1  3055172  3055425      +   \n",
       "318454                    2                        1  3844004  3844212      -   \n",
       "318455                    2                        1  1862773  1863068      -   \n",
       "318456                    1                        1  2900333  2900531      +   \n",
       "\n",
       "        variation  \n",
       "0        0.051998  \n",
       "1        0.261046  \n",
       "2        1.883143  \n",
       "3        0.135708  \n",
       "4        0.142033  \n",
       "...           ...  \n",
       "318452   0.023327  \n",
       "318453   0.733025  \n",
       "318454   0.687423  \n",
       "318455   1.088059  \n",
       "318456   0.417133  \n",
       "\n",
       "[318457 rows x 13 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.read_csv(\"frag-rLP5-M9_expression.txt\", delimiter=\" \")\n",
    "df2\n",
    "# Expression levels from genomic fragment MPRA (random 200–300 bp sheared fragments) in M9 media at rLP5\n",
    "\n",
    "# fragment: 150–300 bp genomic DNA sequence that were randomly sheared and barcoded\n",
    "# RNA_exp_1: RNA expression level (replicate 1) – normalized measurement of transcript abundance for this fragment in the first RNA-Seq replicate. (normalized by DNA)\n",
    "# RNA_exp_2: RNA expression level (replicate 2) – same as above, but from a second biological replicate.\n",
    "# RNA_exp_ave: Average RNA expression – mean of RNA_exp_1 and RNA_exp_2\n",
    "# DNA_sum_1: DNA integration abundance (replicate 1) – quantifies how much of this DNA fragment (or barcode) was actually integrated in the first DNA-Seq sample. \n",
    "# DNA_sum_2: DNA integration abundance (replicate 2) – same as above, from the second replicate.\n",
    "# DNA_ave: Average DNA integration level – mean of DNA_sum_1 and DNA_sum_2\n",
    "# num_integrated_barcodes: Number of barcodes that were integrated into the genome for this fragment. Typically should match num_mapped_barcodes unless there are sequencing/integration issues.\n",
    "# start: Start coordinate (in bp) in the E. coli MG1655 reference genome (U00096.2). \n",
    "# end: End coordinate in the E. coli MG1655 reference genome (U00096.2). \n",
    "# strand: DNA strand orientation (+ or -) \n",
    "# variation: Standard deviation or variability in RNA measurement across barcode replicates \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "73fdb186-fa29-4029-8968-9cbefb0da700",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>tss_name</th>\n",
       "      <th>tss_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>scramble_start</th>\n",
       "      <th>scramble_end</th>\n",
       "      <th>var_left</th>\n",
       "      <th>var_right</th>\n",
       "      <th>scramble_start_pos</th>\n",
       "      <th>scramble_end_pos</th>\n",
       "      <th>scramble_pos_rel_tss</th>\n",
       "      <th>variant</th>\n",
       "      <th>RNA_exp_sum_1_1</th>\n",
       "      <th>RNA_exp_sum_1_2</th>\n",
       "      <th>RNA_exp_sum_2_1</th>\n",
       "      <th>RNA_exp_sum_2_2</th>\n",
       "      <th>RNA_exp_sum_1</th>\n",
       "      <th>RNA_exp_sum_2</th>\n",
       "      <th>RNA_exp_sum_ave</th>\n",
       "      <th>DNA_1</th>\n",
       "      <th>DNA_2</th>\n",
       "      <th>DNA_ave</th>\n",
       "      <th>expn_med</th>\n",
       "      <th>num_barcodes_mapped</th>\n",
       "      <th>num_barcodes_integrated</th>\n",
       "      <th>category</th>\n",
       "      <th>expn_med_fitted</th>\n",
       "      <th>expn_med_fitted_scaled</th>\n",
       "      <th>unscrambled_exp</th>\n",
       "      <th>relative_exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSS_5574_storz,1423414,-_scrambled140-150</td>\n",
       "      <td>TSS_5574_storz</td>\n",
       "      <td>1423414.0</td>\n",
       "      <td>-</td>\n",
       "      <td>140.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>1423384.0</td>\n",
       "      <td>1423534.0</td>\n",
       "      <td>1423524.0</td>\n",
       "      <td>1423534.0</td>\n",
       "      <td>-110.0</td>\n",
       "      <td>GGATTAGTCCAGCCGAATTGCTTAATACTTTCTACCAGTTGTGCCA...</td>\n",
       "      <td>0.504318</td>\n",
       "      <td>0.689500</td>\n",
       "      <td>0.654339</td>\n",
       "      <td>0.636485</td>\n",
       "      <td>0.596909</td>\n",
       "      <td>0.645412</td>\n",
       "      <td>0.621161</td>\n",
       "      <td>36.404637</td>\n",
       "      <td>35.401688</td>\n",
       "      <td>35.903162</td>\n",
       "      <td>0.500461</td>\n",
       "      <td>39</td>\n",
       "      <td>23</td>\n",
       "      <td>scramble</td>\n",
       "      <td>2.539165</td>\n",
       "      <td>0.611763</td>\n",
       "      <td>0.652428</td>\n",
       "      <td>0.952075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TSS_7055_storz_regulondb,1846816,+_scrambled70-80</td>\n",
       "      <td>TSS_7055_storz_regulondb</td>\n",
       "      <td>1846816.0</td>\n",
       "      <td>+</td>\n",
       "      <td>70.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1846696.0</td>\n",
       "      <td>1846846.0</td>\n",
       "      <td>1846766.0</td>\n",
       "      <td>1846776.0</td>\n",
       "      <td>-50.0</td>\n",
       "      <td>CCATTTGATTAACTCCTGTCGTGATATTTATTCACAAAATTAACAC...</td>\n",
       "      <td>0.146131</td>\n",
       "      <td>0.113741</td>\n",
       "      <td>0.182050</td>\n",
       "      <td>0.181035</td>\n",
       "      <td>0.129936</td>\n",
       "      <td>0.181542</td>\n",
       "      <td>0.155739</td>\n",
       "      <td>31.003500</td>\n",
       "      <td>28.704602</td>\n",
       "      <td>29.854051</td>\n",
       "      <td>0.130204</td>\n",
       "      <td>25</td>\n",
       "      <td>12</td>\n",
       "      <td>scramble</td>\n",
       "      <td>1.691925</td>\n",
       "      <td>0.407636</td>\n",
       "      <td>0.258146</td>\n",
       "      <td>0.603299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TSS_17515_regulondb,4477590,-_scrambled135-145</td>\n",
       "      <td>TSS_17515_regulondb</td>\n",
       "      <td>4477590.0</td>\n",
       "      <td>-</td>\n",
       "      <td>135.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>4477560.0</td>\n",
       "      <td>4477710.0</td>\n",
       "      <td>4477695.0</td>\n",
       "      <td>4477705.0</td>\n",
       "      <td>-105.0</td>\n",
       "      <td>ATACAGACAGTAAAATGTTCTGCCACGTTACAATACTCTTATCAGT...</td>\n",
       "      <td>1.110148</td>\n",
       "      <td>1.126184</td>\n",
       "      <td>1.072438</td>\n",
       "      <td>1.179107</td>\n",
       "      <td>1.118166</td>\n",
       "      <td>1.125772</td>\n",
       "      <td>1.121969</td>\n",
       "      <td>15.980134</td>\n",
       "      <td>17.984036</td>\n",
       "      <td>16.982085</td>\n",
       "      <td>0.888817</td>\n",
       "      <td>25</td>\n",
       "      <td>13</td>\n",
       "      <td>scramble</td>\n",
       "      <td>3.427820</td>\n",
       "      <td>0.825867</td>\n",
       "      <td>0.787990</td>\n",
       "      <td>1.423837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TSS_4927_regulondb,1269614,+_scrambled10-20</td>\n",
       "      <td>TSS_4927_regulondb</td>\n",
       "      <td>1269614.0</td>\n",
       "      <td>+</td>\n",
       "      <td>10.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1269494.0</td>\n",
       "      <td>1269644.0</td>\n",
       "      <td>1269504.0</td>\n",
       "      <td>1269514.0</td>\n",
       "      <td>-110.0</td>\n",
       "      <td>GCGGTAATAACCCTGCGATCGATCGGTGCTGCCAGGTCGTGCCAGA...</td>\n",
       "      <td>1.166634</td>\n",
       "      <td>1.650663</td>\n",
       "      <td>1.004031</td>\n",
       "      <td>1.357759</td>\n",
       "      <td>1.408649</td>\n",
       "      <td>1.180895</td>\n",
       "      <td>1.294772</td>\n",
       "      <td>7.739634</td>\n",
       "      <td>5.934540</td>\n",
       "      <td>6.837087</td>\n",
       "      <td>1.061866</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>scramble</td>\n",
       "      <td>3.823799</td>\n",
       "      <td>0.921270</td>\n",
       "      <td>1.344917</td>\n",
       "      <td>0.962715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TSS_17369_storz,4439341,-_scrambled140-150</td>\n",
       "      <td>TSS_17369_storz</td>\n",
       "      <td>4439341.0</td>\n",
       "      <td>-</td>\n",
       "      <td>140.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>4439311.0</td>\n",
       "      <td>4439461.0</td>\n",
       "      <td>4439451.0</td>\n",
       "      <td>4439461.0</td>\n",
       "      <td>-110.0</td>\n",
       "      <td>TATATTGCAAGGACTGCGTAGGCCTGATCGGCATAGCGTATCAGGC...</td>\n",
       "      <td>0.582324</td>\n",
       "      <td>0.917035</td>\n",
       "      <td>0.719130</td>\n",
       "      <td>0.562781</td>\n",
       "      <td>0.749680</td>\n",
       "      <td>0.640956</td>\n",
       "      <td>0.695318</td>\n",
       "      <td>20.349647</td>\n",
       "      <td>20.971240</td>\n",
       "      <td>20.660443</td>\n",
       "      <td>0.767422</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>scramble</td>\n",
       "      <td>3.150038</td>\n",
       "      <td>0.758940</td>\n",
       "      <td>0.656802</td>\n",
       "      <td>1.058641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55312</th>\n",
       "      <td>pos_control_apFAB346</td>\n",
       "      <td>pos_control_apFAB346</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCTTGACAATTAATCATCCGGCTCGTAATGTTTGTGGATAGCT</td>\n",
       "      <td>3.762168</td>\n",
       "      <td>2.852998</td>\n",
       "      <td>2.689369</td>\n",
       "      <td>4.148709</td>\n",
       "      <td>3.307583</td>\n",
       "      <td>3.419039</td>\n",
       "      <td>3.363311</td>\n",
       "      <td>2.017238</td>\n",
       "      <td>1.822945</td>\n",
       "      <td>1.920092</td>\n",
       "      <td>2.581585</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>7.301296</td>\n",
       "      <td>1.759105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55313</th>\n",
       "      <td>pos_control_apFAB87</td>\n",
       "      <td>pos_control_apFAB87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCAAAAAATTTATTTGCTTATTAATCATCCGGCTCGCATA...</td>\n",
       "      <td>1.736385</td>\n",
       "      <td>1.421404</td>\n",
       "      <td>1.646361</td>\n",
       "      <td>1.548851</td>\n",
       "      <td>1.578895</td>\n",
       "      <td>1.597606</td>\n",
       "      <td>1.588251</td>\n",
       "      <td>4.714946</td>\n",
       "      <td>6.376120</td>\n",
       "      <td>5.545533</td>\n",
       "      <td>1.436969</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>4.682128</td>\n",
       "      <td>1.128068</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55314</th>\n",
       "      <td>pos_control_apFAB70</td>\n",
       "      <td>pos_control_apFAB70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCTTGACATCGCATCTTTTTGTACCTATAATGTGTGGATA...</td>\n",
       "      <td>10.670367</td>\n",
       "      <td>13.585706</td>\n",
       "      <td>10.683286</td>\n",
       "      <td>13.121773</td>\n",
       "      <td>12.128037</td>\n",
       "      <td>11.902529</td>\n",
       "      <td>12.015283</td>\n",
       "      <td>7.862836</td>\n",
       "      <td>7.335839</td>\n",
       "      <td>7.599337</td>\n",
       "      <td>11.978279</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>28.803279</td>\n",
       "      <td>6.939591</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55315</th>\n",
       "      <td>pos_control_apFAB49</td>\n",
       "      <td>pos_control_apFAB49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCAAAAAGAGTATTGACTTTTATCCCTTGCGGCGAATACT...</td>\n",
       "      <td>1.135329</td>\n",
       "      <td>1.039307</td>\n",
       "      <td>0.922069</td>\n",
       "      <td>0.733666</td>\n",
       "      <td>1.087318</td>\n",
       "      <td>0.827868</td>\n",
       "      <td>0.957593</td>\n",
       "      <td>3.780015</td>\n",
       "      <td>4.179774</td>\n",
       "      <td>3.979894</td>\n",
       "      <td>0.953658</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>3.576194</td>\n",
       "      <td>0.861615</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55316</th>\n",
       "      <td>pos_control_apFAB44</td>\n",
       "      <td>pos_control_apFAB44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCAAAAAGAGTATTGACTTCGCATCTTTTTGTACCATACT...</td>\n",
       "      <td>8.139306</td>\n",
       "      <td>7.875714</td>\n",
       "      <td>14.247679</td>\n",
       "      <td>22.474147</td>\n",
       "      <td>8.007510</td>\n",
       "      <td>18.360913</td>\n",
       "      <td>13.184211</td>\n",
       "      <td>4.460485</td>\n",
       "      <td>2.079920</td>\n",
       "      <td>3.270202</td>\n",
       "      <td>13.152281</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>31.489688</td>\n",
       "      <td>7.586829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55317 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    name  \\\n",
       "0              TSS_5574_storz,1423414,-_scrambled140-150   \n",
       "1      TSS_7055_storz_regulondb,1846816,+_scrambled70-80   \n",
       "2         TSS_17515_regulondb,4477590,-_scrambled135-145   \n",
       "3            TSS_4927_regulondb,1269614,+_scrambled10-20   \n",
       "4             TSS_17369_storz,4439341,-_scrambled140-150   \n",
       "...                                                  ...   \n",
       "55312                               pos_control_apFAB346   \n",
       "55313                                pos_control_apFAB87   \n",
       "55314                                pos_control_apFAB70   \n",
       "55315                                pos_control_apFAB49   \n",
       "55316                                pos_control_apFAB44   \n",
       "\n",
       "                       tss_name  tss_position strand  scramble_start  \\\n",
       "0                TSS_5574_storz     1423414.0      -           140.0   \n",
       "1      TSS_7055_storz_regulondb     1846816.0      +            70.0   \n",
       "2           TSS_17515_regulondb     4477590.0      -           135.0   \n",
       "3            TSS_4927_regulondb     1269614.0      +            10.0   \n",
       "4               TSS_17369_storz     4439341.0      -           140.0   \n",
       "...                         ...           ...    ...             ...   \n",
       "55312      pos_control_apFAB346           NaN    NaN             NaN   \n",
       "55313       pos_control_apFAB87           NaN    NaN             NaN   \n",
       "55314       pos_control_apFAB70           NaN    NaN             NaN   \n",
       "55315       pos_control_apFAB49           NaN    NaN             NaN   \n",
       "55316       pos_control_apFAB44           NaN    NaN             NaN   \n",
       "\n",
       "       scramble_end   var_left  var_right  scramble_start_pos  \\\n",
       "0             150.0  1423384.0  1423534.0           1423524.0   \n",
       "1              80.0  1846696.0  1846846.0           1846766.0   \n",
       "2             145.0  4477560.0  4477710.0           4477695.0   \n",
       "3              20.0  1269494.0  1269644.0           1269504.0   \n",
       "4             150.0  4439311.0  4439461.0           4439451.0   \n",
       "...             ...        ...        ...                 ...   \n",
       "55312           NaN        NaN        NaN                 NaN   \n",
       "55313           NaN        NaN        NaN                 NaN   \n",
       "55314           NaN        NaN        NaN                 NaN   \n",
       "55315           NaN        NaN        NaN                 NaN   \n",
       "55316           NaN        NaN        NaN                 NaN   \n",
       "\n",
       "       scramble_end_pos  scramble_pos_rel_tss  \\\n",
       "0             1423534.0                -110.0   \n",
       "1             1846776.0                 -50.0   \n",
       "2             4477705.0                -105.0   \n",
       "3             1269514.0                -110.0   \n",
       "4             4439461.0                -110.0   \n",
       "...                 ...                   ...   \n",
       "55312               NaN                   NaN   \n",
       "55313               NaN                   NaN   \n",
       "55314               NaN                   NaN   \n",
       "55315               NaN                   NaN   \n",
       "55316               NaN                   NaN   \n",
       "\n",
       "                                                 variant  RNA_exp_sum_1_1  \\\n",
       "0      GGATTAGTCCAGCCGAATTGCTTAATACTTTCTACCAGTTGTGCCA...         0.504318   \n",
       "1      CCATTTGATTAACTCCTGTCGTGATATTTATTCACAAAATTAACAC...         0.146131   \n",
       "2      ATACAGACAGTAAAATGTTCTGCCACGTTACAATACTCTTATCAGT...         1.110148   \n",
       "3      GCGGTAATAACCCTGCGATCGATCGGTGCTGCCAGGTCGTGCCAGA...         1.166634   \n",
       "4      TATATTGCAAGGACTGCGTAGGCCTGATCGGCATAGCGTATCAGGC...         0.582324   \n",
       "...                                                  ...              ...   \n",
       "55312  GGCGCGCCTTGACAATTAATCATCCGGCTCGTAATGTTTGTGGATAGCT         3.762168   \n",
       "55313  GGCGCGCCAAAAAATTTATTTGCTTATTAATCATCCGGCTCGCATA...         1.736385   \n",
       "55314  GGCGCGCCTTGACATCGCATCTTTTTGTACCTATAATGTGTGGATA...        10.670367   \n",
       "55315  GGCGCGCCAAAAAGAGTATTGACTTTTATCCCTTGCGGCGAATACT...         1.135329   \n",
       "55316  GGCGCGCCAAAAAGAGTATTGACTTCGCATCTTTTTGTACCATACT...         8.139306   \n",
       "\n",
       "       RNA_exp_sum_1_2  RNA_exp_sum_2_1  RNA_exp_sum_2_2  RNA_exp_sum_1  \\\n",
       "0             0.689500         0.654339         0.636485       0.596909   \n",
       "1             0.113741         0.182050         0.181035       0.129936   \n",
       "2             1.126184         1.072438         1.179107       1.118166   \n",
       "3             1.650663         1.004031         1.357759       1.408649   \n",
       "4             0.917035         0.719130         0.562781       0.749680   \n",
       "...                ...              ...              ...            ...   \n",
       "55312         2.852998         2.689369         4.148709       3.307583   \n",
       "55313         1.421404         1.646361         1.548851       1.578895   \n",
       "55314        13.585706        10.683286        13.121773      12.128037   \n",
       "55315         1.039307         0.922069         0.733666       1.087318   \n",
       "55316         7.875714        14.247679        22.474147       8.007510   \n",
       "\n",
       "       RNA_exp_sum_2  RNA_exp_sum_ave      DNA_1      DNA_2    DNA_ave  \\\n",
       "0           0.645412         0.621161  36.404637  35.401688  35.903162   \n",
       "1           0.181542         0.155739  31.003500  28.704602  29.854051   \n",
       "2           1.125772         1.121969  15.980134  17.984036  16.982085   \n",
       "3           1.180895         1.294772   7.739634   5.934540   6.837087   \n",
       "4           0.640956         0.695318  20.349647  20.971240  20.660443   \n",
       "...              ...              ...        ...        ...        ...   \n",
       "55312       3.419039         3.363311   2.017238   1.822945   1.920092   \n",
       "55313       1.597606         1.588251   4.714946   6.376120   5.545533   \n",
       "55314      11.902529        12.015283   7.862836   7.335839   7.599337   \n",
       "55315       0.827868         0.957593   3.780015   4.179774   3.979894   \n",
       "55316      18.360913        13.184211   4.460485   2.079920   3.270202   \n",
       "\n",
       "        expn_med  num_barcodes_mapped  num_barcodes_integrated     category  \\\n",
       "0       0.500461                   39                       23     scramble   \n",
       "1       0.130204                   25                       12     scramble   \n",
       "2       0.888817                   25                       13     scramble   \n",
       "3       1.061866                   15                        6     scramble   \n",
       "4       0.767422                   18                        9     scramble   \n",
       "...          ...                  ...                      ...          ...   \n",
       "55312   2.581585                    5                        3  pos_control   \n",
       "55313   1.436969                    7                        4  pos_control   \n",
       "55314  11.978279                    6                        3  pos_control   \n",
       "55315   0.953658                    5                        3  pos_control   \n",
       "55316  13.152281                    5                        3  pos_control   \n",
       "\n",
       "       expn_med_fitted  expn_med_fitted_scaled  unscrambled_exp  relative_exp  \n",
       "0             2.539165                0.611763         0.652428      0.952075  \n",
       "1             1.691925                0.407636         0.258146      0.603299  \n",
       "2             3.427820                0.825867         0.787990      1.423837  \n",
       "3             3.823799                0.921270         1.344917      0.962715  \n",
       "4             3.150038                0.758940         0.656802      1.058641  \n",
       "...                ...                     ...              ...           ...  \n",
       "55312         7.301296                1.759105              NaN           NaN  \n",
       "55313         4.682128                1.128068              NaN           NaN  \n",
       "55314        28.803279                6.939591              NaN           NaN  \n",
       "55315         3.576194                0.861615              NaN           NaN  \n",
       "55316        31.489688                7.586829              NaN           NaN  \n",
       "\n",
       "[55317 rows x 30 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df4 = pd.read_csv(\"endo_scramble_expression_formatted_std.txt\", delimiter= \"\\t\")\n",
    "df4\n",
    "# Positive promoter controls are synthetic thus have no actual location.\n",
    "# This refers to the 2,000 out of the 17,000 TSS they found, scrambled them to find the functional sites.\n",
    "# The dataset includes the controls, scrambled variants, and unscrambled\n",
    "\n",
    "# name: ID represented as {[TSS name][genomic position][strand + the scrambled region]}\n",
    "# tss_name: Name of the original (unscrambled) transcription start site (TSS) this variant is derived from.\n",
    "# tss_position: Start genome coordinate of the TSS.\n",
    "# strand: Strand direction\n",
    "# scramble_start: Start of the 10 bp scrambled window relative to var_left and var_right\n",
    "# scramble_end: End of the scrambled window relative to var_left and var_right\n",
    "# var_left: Genomic coordinate of the start (5' end) of the full 150 bp promoter variant.\n",
    "# var_right: Genomic coordinate of the end (3' end) of the full 150 bp promoter variant.\n",
    "# scramble_start_pos: Genomic coordinate where the scrambled 10 bp region begins.\n",
    "# scramble_end_pos: Genomic coordinate where the scrambled 10 bp region ends.\n",
    "# scramble_pos_rel_tss: Position of the scrambled region relative to the TSS.\n",
    "# variant: The sequence\n",
    "# RNA_exp_sum_1_1: Sum of RNA expression counts for barcodes in replicate 1, technical replicate 1 (normalized by DNA?)\n",
    "# RNA_exp_sum_1_2: Sum of RNA expression counts for barcodes in replicate 1, technical replicate 2.\n",
    "# RNA_exp_sum_2_1: Sum of RNA expression counts for barcodes in replicate 2, technical replicate 1.\n",
    "# RNA_exp_sum_2_2: Sum of RNA expression counts for barcodes in replicate 2, technical replicate 2.\n",
    "# RNA_exp_sum_1: Sum of RNA expression across both technical replicates for biological replicate 1.\n",
    "# RNA_exp_sum_2: Sum of RNA expression across both technical replicates for biological replicate 2.\n",
    "# RNA_exp_sum_ave: (RNA_exp_sum_1 + RNA_exp_sum_2)/2)\n",
    "# DNA_1: DNA-seq count in biological replicate 1 (how much of this fragment was integrated).\n",
    "# DNA_2: DNA-seq count in biological replicate 2\n",
    "# DNA_ave: (DNA_1 + DNA_2) / 2\n",
    "# expn_med: Median RNA expression across all barcodes, normalized by DNA abundance (RNA/DNA).\n",
    "# num_barcodes_integrated: Number of barcodes actually integrated and measured in DNA/RNA-seq.\n",
    "# category: Label indicating whether this is scrambled, unscrambled, negative, positive control\n",
    "# unscrambled_exp: Measured promoter activity of the wild-type (unscrambled) version of this promoter.\n",
    "# relative_exp: (expn_med / unscrambled_exp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d391cd4e-cbcb-422e-a861-ad204456ac0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>tss_name</th>\n",
       "      <th>tss_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>variant</th>\n",
       "      <th>RNA_exp_sum_1_1</th>\n",
       "      <th>RNA_exp_sum_1_2</th>\n",
       "      <th>RNA_exp_sum_2_1</th>\n",
       "      <th>RNA_exp_sum_2_2</th>\n",
       "      <th>RNA_exp_sum_1</th>\n",
       "      <th>RNA_exp_sum_2</th>\n",
       "      <th>RNA_exp_sum_ave</th>\n",
       "      <th>DNA_1</th>\n",
       "      <th>DNA_2</th>\n",
       "      <th>DNA_ave</th>\n",
       "      <th>expn_med</th>\n",
       "      <th>num_barcodes_mapped</th>\n",
       "      <th>num_barcodes_integrated</th>\n",
       "      <th>category</th>\n",
       "      <th>expn_med_fitted</th>\n",
       "      <th>expn_med_fitted_scaled</th>\n",
       "      <th>active</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSS_11125_storz_regulondb,2945404,+</td>\n",
       "      <td>TSS_11125_storz_regulondb</td>\n",
       "      <td>2945404.0</td>\n",
       "      <td>+</td>\n",
       "      <td>2945284.0</td>\n",
       "      <td>2945434.0</td>\n",
       "      <td>GCAAATTTTGCACAAAAAATAGGCTTTAGTGATTTGTTTTTGTTCA...</td>\n",
       "      <td>43.807236</td>\n",
       "      <td>55.833039</td>\n",
       "      <td>54.661823</td>\n",
       "      <td>49.577995</td>\n",
       "      <td>49.820137</td>\n",
       "      <td>52.119909</td>\n",
       "      <td>50.970023</td>\n",
       "      <td>31.625906</td>\n",
       "      <td>35.101752</td>\n",
       "      <td>33.363829</td>\n",
       "      <td>52.010547</td>\n",
       "      <td>50</td>\n",
       "      <td>28</td>\n",
       "      <td>tss</td>\n",
       "      <td>46.359818</td>\n",
       "      <td>46.076834</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TSS_1218_storz,289782,+</td>\n",
       "      <td>TSS_1218_storz</td>\n",
       "      <td>289782.0</td>\n",
       "      <td>+</td>\n",
       "      <td>289662.0</td>\n",
       "      <td>289812.0</td>\n",
       "      <td>GTATCTGCCTCCGATTCTCTGCAGAAGCAGAAAGACATTGGATCGA...</td>\n",
       "      <td>0.562627</td>\n",
       "      <td>0.549348</td>\n",
       "      <td>0.816074</td>\n",
       "      <td>0.585183</td>\n",
       "      <td>0.555988</td>\n",
       "      <td>0.700629</td>\n",
       "      <td>0.628308</td>\n",
       "      <td>36.298833</td>\n",
       "      <td>31.606403</td>\n",
       "      <td>33.952618</td>\n",
       "      <td>0.565387</td>\n",
       "      <td>53</td>\n",
       "      <td>26</td>\n",
       "      <td>tss</td>\n",
       "      <td>0.749499</td>\n",
       "      <td>0.744924</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TSS_12352_storz,3253602,-</td>\n",
       "      <td>TSS_12352_storz</td>\n",
       "      <td>3253602.0</td>\n",
       "      <td>-</td>\n",
       "      <td>3253572.0</td>\n",
       "      <td>3253722.0</td>\n",
       "      <td>CGTTTGTCTGCGCTGTGTGCCGCAACGACCGCAGCAATGGGGGCCG...</td>\n",
       "      <td>0.516503</td>\n",
       "      <td>0.717569</td>\n",
       "      <td>0.658121</td>\n",
       "      <td>0.587724</td>\n",
       "      <td>0.617036</td>\n",
       "      <td>0.622923</td>\n",
       "      <td>0.619979</td>\n",
       "      <td>72.105393</td>\n",
       "      <td>72.498105</td>\n",
       "      <td>72.301749</td>\n",
       "      <td>0.577755</td>\n",
       "      <td>108</td>\n",
       "      <td>54</td>\n",
       "      <td>tss</td>\n",
       "      <td>0.760464</td>\n",
       "      <td>0.755822</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TSS_6478_storz,1684674,-</td>\n",
       "      <td>TSS_6478_storz</td>\n",
       "      <td>1684674.0</td>\n",
       "      <td>-</td>\n",
       "      <td>1684644.0</td>\n",
       "      <td>1684794.0</td>\n",
       "      <td>GCAGATACAACTCACACAATGCACCCGCTGTGTGAAATAAACAGAG...</td>\n",
       "      <td>1.006087</td>\n",
       "      <td>1.084363</td>\n",
       "      <td>0.763596</td>\n",
       "      <td>0.900281</td>\n",
       "      <td>1.045225</td>\n",
       "      <td>0.831939</td>\n",
       "      <td>0.938582</td>\n",
       "      <td>35.065895</td>\n",
       "      <td>37.408634</td>\n",
       "      <td>36.237265</td>\n",
       "      <td>0.936175</td>\n",
       "      <td>56</td>\n",
       "      <td>27</td>\n",
       "      <td>tss</td>\n",
       "      <td>1.078232</td>\n",
       "      <td>1.071651</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TSS_2956_storz,770940,+</td>\n",
       "      <td>TSS_2956_storz</td>\n",
       "      <td>770940.0</td>\n",
       "      <td>+</td>\n",
       "      <td>770820.0</td>\n",
       "      <td>770970.0</td>\n",
       "      <td>TTATAAAGATATGACCAAGTTCTGGGGCAAGTTGTTTGGTATCAAC...</td>\n",
       "      <td>0.715541</td>\n",
       "      <td>0.744280</td>\n",
       "      <td>0.679240</td>\n",
       "      <td>0.634099</td>\n",
       "      <td>0.729911</td>\n",
       "      <td>0.656670</td>\n",
       "      <td>0.693290</td>\n",
       "      <td>79.950519</td>\n",
       "      <td>78.564136</td>\n",
       "      <td>79.257327</td>\n",
       "      <td>0.632548</td>\n",
       "      <td>116</td>\n",
       "      <td>59</td>\n",
       "      <td>tss</td>\n",
       "      <td>0.809042</td>\n",
       "      <td>0.804104</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17616</th>\n",
       "      <td>pos_control_apFAB54</td>\n",
       "      <td>pos_control_apFAB54</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCTTGACATAAAGTCTAACCTATAGGATACTTACAGCCAT...</td>\n",
       "      <td>38.494902</td>\n",
       "      <td>57.975130</td>\n",
       "      <td>39.317072</td>\n",
       "      <td>44.515699</td>\n",
       "      <td>48.235016</td>\n",
       "      <td>41.916385</td>\n",
       "      <td>45.075701</td>\n",
       "      <td>3.128893</td>\n",
       "      <td>3.379367</td>\n",
       "      <td>3.254130</td>\n",
       "      <td>45.188710</td>\n",
       "      <td>28</td>\n",
       "      <td>5</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>40.311705</td>\n",
       "      <td>40.065639</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17617</th>\n",
       "      <td>pos_control_apFAB36</td>\n",
       "      <td>pos_control_apFAB36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCAAAAAGAGTATTGACTATTAATCATCCGGCTCGTATAA...</td>\n",
       "      <td>58.847253</td>\n",
       "      <td>40.995121</td>\n",
       "      <td>61.542591</td>\n",
       "      <td>50.045329</td>\n",
       "      <td>49.921187</td>\n",
       "      <td>55.793960</td>\n",
       "      <td>52.857573</td>\n",
       "      <td>2.569912</td>\n",
       "      <td>3.470787</td>\n",
       "      <td>3.020350</td>\n",
       "      <td>54.347974</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>48.432137</td>\n",
       "      <td>48.136503</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17618</th>\n",
       "      <td>pos_control_apFAB82</td>\n",
       "      <td>pos_control_apFAB82</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCAAAAAATTTATTTGCTTTAAAGTCTAACCTATAGGCAT...</td>\n",
       "      <td>8.342111</td>\n",
       "      <td>7.818100</td>\n",
       "      <td>5.151892</td>\n",
       "      <td>7.300009</td>\n",
       "      <td>8.080105</td>\n",
       "      <td>6.225951</td>\n",
       "      <td>7.153028</td>\n",
       "      <td>5.793876</td>\n",
       "      <td>8.068636</td>\n",
       "      <td>6.931256</td>\n",
       "      <td>7.260103</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>6.684909</td>\n",
       "      <td>6.644103</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17619</th>\n",
       "      <td>pos_control_apFAB114</td>\n",
       "      <td>pos_control_apFAB114</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCTCGACATCAGGAAAATTTTTCTGATACTTACAGCCATG...</td>\n",
       "      <td>0.696238</td>\n",
       "      <td>0.544310</td>\n",
       "      <td>0.953141</td>\n",
       "      <td>0.916195</td>\n",
       "      <td>0.620274</td>\n",
       "      <td>0.934668</td>\n",
       "      <td>0.777471</td>\n",
       "      <td>11.040219</td>\n",
       "      <td>7.347735</td>\n",
       "      <td>9.193977</td>\n",
       "      <td>0.798860</td>\n",
       "      <td>20</td>\n",
       "      <td>8</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>0.956492</td>\n",
       "      <td>0.950653</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17620</th>\n",
       "      <td>pos_control_apFAB53</td>\n",
       "      <td>pos_control_apFAB53</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GGCGCGCCAAAAAGAGTATTGACTTTTATCCCTTGCGGCGATAGAT...</td>\n",
       "      <td>53.492610</td>\n",
       "      <td>121.696207</td>\n",
       "      <td>91.054765</td>\n",
       "      <td>52.409801</td>\n",
       "      <td>87.594409</td>\n",
       "      <td>71.732283</td>\n",
       "      <td>79.663346</td>\n",
       "      <td>1.918849</td>\n",
       "      <td>3.391648</td>\n",
       "      <td>2.655248</td>\n",
       "      <td>80.983733</td>\n",
       "      <td>29</td>\n",
       "      <td>3</td>\n",
       "      <td>pos_control</td>\n",
       "      <td>72.046904</td>\n",
       "      <td>71.607124</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17621 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      name                   tss_name  \\\n",
       "0      TSS_11125_storz_regulondb,2945404,+  TSS_11125_storz_regulondb   \n",
       "1                  TSS_1218_storz,289782,+             TSS_1218_storz   \n",
       "2                TSS_12352_storz,3253602,-            TSS_12352_storz   \n",
       "3                 TSS_6478_storz,1684674,-             TSS_6478_storz   \n",
       "4                  TSS_2956_storz,770940,+             TSS_2956_storz   \n",
       "...                                    ...                        ...   \n",
       "17616                  pos_control_apFAB54        pos_control_apFAB54   \n",
       "17617                  pos_control_apFAB36        pos_control_apFAB36   \n",
       "17618                  pos_control_apFAB82        pos_control_apFAB82   \n",
       "17619                 pos_control_apFAB114       pos_control_apFAB114   \n",
       "17620                  pos_control_apFAB53        pos_control_apFAB53   \n",
       "\n",
       "       tss_position strand      start        end  \\\n",
       "0         2945404.0      +  2945284.0  2945434.0   \n",
       "1          289782.0      +   289662.0   289812.0   \n",
       "2         3253602.0      -  3253572.0  3253722.0   \n",
       "3         1684674.0      -  1684644.0  1684794.0   \n",
       "4          770940.0      +   770820.0   770970.0   \n",
       "...             ...    ...        ...        ...   \n",
       "17616           NaN    NaN        NaN        NaN   \n",
       "17617           NaN    NaN        NaN        NaN   \n",
       "17618           NaN    NaN        NaN        NaN   \n",
       "17619           NaN    NaN        NaN        NaN   \n",
       "17620           NaN    NaN        NaN        NaN   \n",
       "\n",
       "                                                 variant  RNA_exp_sum_1_1  \\\n",
       "0      GCAAATTTTGCACAAAAAATAGGCTTTAGTGATTTGTTTTTGTTCA...        43.807236   \n",
       "1      GTATCTGCCTCCGATTCTCTGCAGAAGCAGAAAGACATTGGATCGA...         0.562627   \n",
       "2      CGTTTGTCTGCGCTGTGTGCCGCAACGACCGCAGCAATGGGGGCCG...         0.516503   \n",
       "3      GCAGATACAACTCACACAATGCACCCGCTGTGTGAAATAAACAGAG...         1.006087   \n",
       "4      TTATAAAGATATGACCAAGTTCTGGGGCAAGTTGTTTGGTATCAAC...         0.715541   \n",
       "...                                                  ...              ...   \n",
       "17616  GGCGCGCCTTGACATAAAGTCTAACCTATAGGATACTTACAGCCAT...        38.494902   \n",
       "17617  GGCGCGCCAAAAAGAGTATTGACTATTAATCATCCGGCTCGTATAA...        58.847253   \n",
       "17618  GGCGCGCCAAAAAATTTATTTGCTTTAAAGTCTAACCTATAGGCAT...         8.342111   \n",
       "17619  GGCGCGCCTCGACATCAGGAAAATTTTTCTGATACTTACAGCCATG...         0.696238   \n",
       "17620  GGCGCGCCAAAAAGAGTATTGACTTTTATCCCTTGCGGCGATAGAT...        53.492610   \n",
       "\n",
       "       RNA_exp_sum_1_2  RNA_exp_sum_2_1  RNA_exp_sum_2_2  RNA_exp_sum_1  \\\n",
       "0            55.833039        54.661823        49.577995      49.820137   \n",
       "1             0.549348         0.816074         0.585183       0.555988   \n",
       "2             0.717569         0.658121         0.587724       0.617036   \n",
       "3             1.084363         0.763596         0.900281       1.045225   \n",
       "4             0.744280         0.679240         0.634099       0.729911   \n",
       "...                ...              ...              ...            ...   \n",
       "17616        57.975130        39.317072        44.515699      48.235016   \n",
       "17617        40.995121        61.542591        50.045329      49.921187   \n",
       "17618         7.818100         5.151892         7.300009       8.080105   \n",
       "17619         0.544310         0.953141         0.916195       0.620274   \n",
       "17620       121.696207        91.054765        52.409801      87.594409   \n",
       "\n",
       "       RNA_exp_sum_2  RNA_exp_sum_ave      DNA_1      DNA_2    DNA_ave  \\\n",
       "0          52.119909        50.970023  31.625906  35.101752  33.363829   \n",
       "1           0.700629         0.628308  36.298833  31.606403  33.952618   \n",
       "2           0.622923         0.619979  72.105393  72.498105  72.301749   \n",
       "3           0.831939         0.938582  35.065895  37.408634  36.237265   \n",
       "4           0.656670         0.693290  79.950519  78.564136  79.257327   \n",
       "...              ...              ...        ...        ...        ...   \n",
       "17616      41.916385        45.075701   3.128893   3.379367   3.254130   \n",
       "17617      55.793960        52.857573   2.569912   3.470787   3.020350   \n",
       "17618       6.225951         7.153028   5.793876   8.068636   6.931256   \n",
       "17619       0.934668         0.777471  11.040219   7.347735   9.193977   \n",
       "17620      71.732283        79.663346   1.918849   3.391648   2.655248   \n",
       "\n",
       "        expn_med  num_barcodes_mapped  num_barcodes_integrated     category  \\\n",
       "0      52.010547                   50                       28          tss   \n",
       "1       0.565387                   53                       26          tss   \n",
       "2       0.577755                  108                       54          tss   \n",
       "3       0.936175                   56                       27          tss   \n",
       "4       0.632548                  116                       59          tss   \n",
       "...          ...                  ...                      ...          ...   \n",
       "17616  45.188710                   28                        5  pos_control   \n",
       "17617  54.347974                   14                        4  pos_control   \n",
       "17618   7.260103                   13                        5  pos_control   \n",
       "17619   0.798860                   20                        8  pos_control   \n",
       "17620  80.983733                   29                        3  pos_control   \n",
       "\n",
       "       expn_med_fitted  expn_med_fitted_scaled    active  \n",
       "0            46.359818               46.076834    active  \n",
       "1             0.749499                0.744924  inactive  \n",
       "2             0.760464                0.755822  inactive  \n",
       "3             1.078232                1.071651    active  \n",
       "4             0.809042                0.804104  inactive  \n",
       "...                ...                     ...       ...  \n",
       "17616        40.311705               40.065639    active  \n",
       "17617        48.432137               48.136503    active  \n",
       "17618         6.684909                6.644103    active  \n",
       "17619         0.956492                0.950653  inactive  \n",
       "17620        72.046904               71.607124    active  \n",
       "\n",
       "[17621 rows x 24 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df5 = pd.read_csv(\"fLP3_Endo2_lb_expression_formatted_std.txt\", delimiter=\"\\t\")\n",
    "df5\n",
    "# Expression summary for the TSS promoter library (17,635 reported TSSs) in LB, integrated at the fLP3 site\n",
    "# name: Unique identifier for the promoter variant in the format [TSS_name,TSS_position,strand]\n",
    "# tss_name: The named identifier for the transcription start site (TSS), such as from RegulonDB or Storz \n",
    "# tss_position: Genomic coordinate of the TSS — the position where transcription starts.\n",
    "# strand: Strand direction\n",
    "# start: Genomic coordinate where the 150 bp promoter fragment starts.\n",
    "# end: Genomic coordinate where the 150 bp promoter fragment ends.\n",
    "# variant: The promoter sequence\n",
    "# RNA_exp_sum_1_1: Sum of RNA-seq counts for this variant in biological replicate 1, technical replicate 1.\n",
    "# RNA_exp_sum_1_2: Sum of RNA-seq counts in biological replicate 1, technical replicate 2.\n",
    "# RNA_exp_sum_2_1: Sum of RNA-seq counts in biological replicate 2, technical replicate 1.\n",
    "# RNA_exp_sum_2_2: Sum of RNA-seq counts in biological replicate 2, technical replicate 2.\n",
    "# RNA_exp_sum_1: Total RNA expression for biological replicate 1 \n",
    "# RNA_exp_sum_2: Total RNA expression for biological replicate 2 \n",
    "# RNA_exp_sum_ave: Average of RNA expression across the two biological replicates.\n",
    "# DNA_1: DNA-seq count in biological replicate 1 \n",
    "# DNA_2: DNA-seq count in biological replicate 2.\n",
    "# DNA_ave: Average of DNA_1 and DNA_2\n",
    "# expn_med: Median RNA/DNA expression across barcodes \n",
    "# num_barcodes_integrated: Number of barcodes that were actually integrated and sequenced.\n",
    "# category: Type of promoter: ['tss', 'neg_control', 'pos_control']\n",
    "# active: Binary classification — \"active\" or \"inactive\" promoter based on expn_med\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "43d93c2d-08b9-4e49-9497-3063d42a5d52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['tile', 'neg_control', 'random', 'pos_control'], dtype=object)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df6 = pd.read_csv(\"peak_tile_expression_formatted_std.txt\", delimiter=\"\\t\")\n",
    "df6[\"category\"].unique()\n",
    "# The authors first used sheared genomic fragments (~200–300 bp) to find candidate promoter regions, \n",
    "# and then designed a tiling oligo library: ~150 bp sequences overlapping by 10 bp Spanning all ~3,500 candidate promoter regions\n",
    "# LB and nth-ydgR intergenic locus \n",
    "\n",
    "# variant: The 150 bp DNA sequence (oligo) used in the MPRA tile\n",
    "# name: Identifier for the tile, usually formatted as “peak_start_peak_end_strand_posStart-posEnd”.\n",
    "# peak_start: Genomic start coordinate of the candidate promoter region (the whole peak region).\n",
    "# peak_end: Genomic end coordinate of the candidate promoter region.\n",
    "# strand: Strand direction\n",
    "# tile_start: Start of this 150 bp oligo relative to the peak\n",
    "# tile_end: End of the tile \n",
    "# RNA_exp_sum_1_1: Sum of RNA reads for this tile in biological replicate 1, technical replicate 1\n",
    "# RNA_exp_sum_1_2: Same as above, but technical replicate 2\n",
    "# RNA_exp_sum_2_1: Sum of RNA reads in biological replicate 2, technical replicate 1.\n",
    "# RNA_exp_sum_2_2: Same as above, but technical replicate 2.\n",
    "# RNA_exp_sum_1: Total RNA expression in biological replicate 1\n",
    "# RNA_exp_sum_2: Total RNA expression in biological replicate 2 \n",
    "# RNA_exp_sum_ave: Average RNA expression across the two biological replicates.\n",
    "# DNA_1: DNA read count in biological replicate 1\n",
    "# DNA_2: DNA read count in biological replicate 2.\n",
    "# DNA_ave: Average of DNA_1 and DNA_2\n",
    "# expn_med: Median RNA/DNA expression across all barcodes for this tile \n",
    "# num_barcodes_integrated: Number of barcodes that were successfully integrated and sequenced.\n",
    "# category: ['tile', 'neg_control', 'random', 'pos_control']\n",
    "# peak_length: Length (in bp) of the peak region from which this tile was derived (e.g., 362 bp).\n",
    "# tile_start_relative: Position of the tile’s start within the peak (tile_start / peak_length)\n",
    "# start: Genomic start coordinate of the tile (based on tile_start + peak_start).\n",
    "# end: Genomic end coordinate of the tile.\n",
    "# active: Binary label indicating if the tile is active or inactive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3d80c50c-35eb-4be1-8bd4-a249cf1eaa3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>tss_name</th>\n",
       "      <th>tss_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>variant</th>\n",
       "      <th>RNA_exp_1</th>\n",
       "      <th>RNA_exp_2</th>\n",
       "      <th>RNA_exp_sum_ave</th>\n",
       "      <th>DNA_sum</th>\n",
       "      <th>expn_med1</th>\n",
       "      <th>expn_med2</th>\n",
       "      <th>expn_med</th>\n",
       "      <th>num_barcodes_mapped</th>\n",
       "      <th>num_barcodes_integrated</th>\n",
       "      <th>category</th>\n",
       "      <th>expn_med_fitted</th>\n",
       "      <th>expn_med_fitted_scaled</th>\n",
       "      <th>active</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSS_11125_storz_regulondb,2945404,+</td>\n",
       "      <td>TSS_11125_storz_regulondb</td>\n",
       "      <td>2945404.0</td>\n",
       "      <td>+</td>\n",
       "      <td>2945284.0</td>\n",
       "      <td>2945434.0</td>\n",
       "      <td>GCAAATTTTGCACAAAAAATAGGCTTTAGTGATTTGTTTTTGTTCA...</td>\n",
       "      <td>54.590828</td>\n",
       "      <td>53.711399</td>\n",
       "      <td>54.151114</td>\n",
       "      <td>49.578588</td>\n",
       "      <td>55.212763</td>\n",
       "      <td>51.214980</td>\n",
       "      <td>53.213872</td>\n",
       "      <td>60</td>\n",
       "      <td>47</td>\n",
       "      <td>tss</td>\n",
       "      <td>53.213872</td>\n",
       "      <td>60.237193</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TSS_15364_storz,3953827,-</td>\n",
       "      <td>TSS_15364_storz</td>\n",
       "      <td>3953827.0</td>\n",
       "      <td>-</td>\n",
       "      <td>3953797.0</td>\n",
       "      <td>3953947.0</td>\n",
       "      <td>CACCGCCACGCCAGCAGCCAGACCGCCGCCGCCGACTGGCACAAAT...</td>\n",
       "      <td>0.748428</td>\n",
       "      <td>0.669080</td>\n",
       "      <td>0.708754</td>\n",
       "      <td>113.905547</td>\n",
       "      <td>0.611601</td>\n",
       "      <td>0.666221</td>\n",
       "      <td>0.638911</td>\n",
       "      <td>136</td>\n",
       "      <td>106</td>\n",
       "      <td>tss</td>\n",
       "      <td>0.638911</td>\n",
       "      <td>0.723236</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TSS_12352_storz,3253602,-</td>\n",
       "      <td>TSS_12352_storz</td>\n",
       "      <td>3253602.0</td>\n",
       "      <td>-</td>\n",
       "      <td>3253572.0</td>\n",
       "      <td>3253722.0</td>\n",
       "      <td>CGTTTGTCTGCGCTGTGTGCCGCAACGACCGCAGCAATGGGGGCCG...</td>\n",
       "      <td>0.706336</td>\n",
       "      <td>0.711005</td>\n",
       "      <td>0.708670</td>\n",
       "      <td>89.722359</td>\n",
       "      <td>0.639401</td>\n",
       "      <td>0.644214</td>\n",
       "      <td>0.641807</td>\n",
       "      <td>121</td>\n",
       "      <td>100</td>\n",
       "      <td>tss</td>\n",
       "      <td>0.641807</td>\n",
       "      <td>0.726515</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TSS_6478_storz,1684674,-</td>\n",
       "      <td>TSS_6478_storz</td>\n",
       "      <td>1684674.0</td>\n",
       "      <td>-</td>\n",
       "      <td>1684644.0</td>\n",
       "      <td>1684794.0</td>\n",
       "      <td>GCAGATACAACTCACACAATGCACCCGCTGTGTGAAATAAACAGAG...</td>\n",
       "      <td>1.061173</td>\n",
       "      <td>0.855839</td>\n",
       "      <td>0.958506</td>\n",
       "      <td>54.485220</td>\n",
       "      <td>0.978561</td>\n",
       "      <td>0.885794</td>\n",
       "      <td>0.932177</td>\n",
       "      <td>59</td>\n",
       "      <td>52</td>\n",
       "      <td>tss</td>\n",
       "      <td>0.932177</td>\n",
       "      <td>1.055209</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TSS_16748_storz,4275469,+</td>\n",
       "      <td>TSS_16748_storz</td>\n",
       "      <td>4275469.0</td>\n",
       "      <td>+</td>\n",
       "      <td>4275349.0</td>\n",
       "      <td>4275499.0</td>\n",
       "      <td>GTCAATATGCTCGTCAATCCATGCGATAAGATCCTGAATAATTTTC...</td>\n",
       "      <td>0.892963</td>\n",
       "      <td>0.802813</td>\n",
       "      <td>0.847888</td>\n",
       "      <td>48.016955</td>\n",
       "      <td>0.778401</td>\n",
       "      <td>0.718905</td>\n",
       "      <td>0.748653</td>\n",
       "      <td>49</td>\n",
       "      <td>41</td>\n",
       "      <td>tss</td>\n",
       "      <td>0.748653</td>\n",
       "      <td>0.847462</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17762</th>\n",
       "      <td>neg_control_1215677:1215827</td>\n",
       "      <td>neg_control_1215677:1215827</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>1215677.0</td>\n",
       "      <td>1215827.0</td>\n",
       "      <td>AGAATCAGCAGTGTTAGGGCAGGCTGTCACCAATTTAATGCTTTCA...</td>\n",
       "      <td>1.332975</td>\n",
       "      <td>1.417616</td>\n",
       "      <td>1.375296</td>\n",
       "      <td>23.106246</td>\n",
       "      <td>1.223201</td>\n",
       "      <td>1.279809</td>\n",
       "      <td>1.251505</td>\n",
       "      <td>41</td>\n",
       "      <td>27</td>\n",
       "      <td>neg_control</td>\n",
       "      <td>1.251505</td>\n",
       "      <td>1.416682</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17763</th>\n",
       "      <td>neg_control_3720155:3720305</td>\n",
       "      <td>neg_control_3720155:3720305</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>3720155.0</td>\n",
       "      <td>3720305.0</td>\n",
       "      <td>AGACAAACAGTCTCAAGCACCCGTGGCTATTCTAGCTTAATAAGTT...</td>\n",
       "      <td>0.921965</td>\n",
       "      <td>0.881224</td>\n",
       "      <td>0.901595</td>\n",
       "      <td>28.526390</td>\n",
       "      <td>0.856241</td>\n",
       "      <td>0.924306</td>\n",
       "      <td>0.890274</td>\n",
       "      <td>27</td>\n",
       "      <td>23</td>\n",
       "      <td>neg_control</td>\n",
       "      <td>0.890274</td>\n",
       "      <td>1.007774</td>\n",
       "      <td>active</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17764</th>\n",
       "      <td>neg_control_4474820:4474970</td>\n",
       "      <td>neg_control_4474820:4474970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>4474820.0</td>\n",
       "      <td>4474970.0</td>\n",
       "      <td>CTTAAGTAATCGAGAAAAAACAAATTTAATACAAAGGCTATTTGAT...</td>\n",
       "      <td>0.469512</td>\n",
       "      <td>0.601874</td>\n",
       "      <td>0.535693</td>\n",
       "      <td>10.462048</td>\n",
       "      <td>0.407734</td>\n",
       "      <td>0.369723</td>\n",
       "      <td>0.388728</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>neg_control</td>\n",
       "      <td>0.388728</td>\n",
       "      <td>0.440034</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17765</th>\n",
       "      <td>neg_control_533510:533660</td>\n",
       "      <td>neg_control_533510:533660</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>533510.0</td>\n",
       "      <td>533660.0</td>\n",
       "      <td>ACCGGTCAGCAAAATGGCGGTTACAGTTCGTGAAGCGGCGCTGGTG...</td>\n",
       "      <td>0.889601</td>\n",
       "      <td>1.078358</td>\n",
       "      <td>0.983979</td>\n",
       "      <td>0.934222</td>\n",
       "      <td>1.070301</td>\n",
       "      <td>0.693230</td>\n",
       "      <td>0.881765</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neg_control</td>\n",
       "      <td>0.881765</td>\n",
       "      <td>0.998143</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17766</th>\n",
       "      <td>neg_control_4435071:4435221</td>\n",
       "      <td>neg_control_4435071:4435221</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>4435071.0</td>\n",
       "      <td>4435221.0</td>\n",
       "      <td>ACCGTTAACATTGCGCTCATTGACCATGGCAAACCGATTTTAGGCG...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.924306</td>\n",
       "      <td>0.462153</td>\n",
       "      <td>0.506342</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.924306</td>\n",
       "      <td>0.462153</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>neg_control</td>\n",
       "      <td>0.462153</td>\n",
       "      <td>0.523150</td>\n",
       "      <td>inactive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17767 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      name                     tss_name  \\\n",
       "0      TSS_11125_storz_regulondb,2945404,+    TSS_11125_storz_regulondb   \n",
       "1                TSS_15364_storz,3953827,-              TSS_15364_storz   \n",
       "2                TSS_12352_storz,3253602,-              TSS_12352_storz   \n",
       "3                 TSS_6478_storz,1684674,-               TSS_6478_storz   \n",
       "4                TSS_16748_storz,4275469,+              TSS_16748_storz   \n",
       "...                                    ...                          ...   \n",
       "17762          neg_control_1215677:1215827  neg_control_1215677:1215827   \n",
       "17763          neg_control_3720155:3720305  neg_control_3720155:3720305   \n",
       "17764          neg_control_4474820:4474970  neg_control_4474820:4474970   \n",
       "17765            neg_control_533510:533660    neg_control_533510:533660   \n",
       "17766          neg_control_4435071:4435221  neg_control_4435071:4435221   \n",
       "\n",
       "       tss_position strand      start        end  \\\n",
       "0         2945404.0      +  2945284.0  2945434.0   \n",
       "1         3953827.0      -  3953797.0  3953947.0   \n",
       "2         3253602.0      -  3253572.0  3253722.0   \n",
       "3         1684674.0      -  1684644.0  1684794.0   \n",
       "4         4275469.0      +  4275349.0  4275499.0   \n",
       "...             ...    ...        ...        ...   \n",
       "17762           NaN      +  1215677.0  1215827.0   \n",
       "17763           NaN      +  3720155.0  3720305.0   \n",
       "17764           NaN      +  4474820.0  4474970.0   \n",
       "17765           NaN      +   533510.0   533660.0   \n",
       "17766           NaN      +  4435071.0  4435221.0   \n",
       "\n",
       "                                                 variant  RNA_exp_1  \\\n",
       "0      GCAAATTTTGCACAAAAAATAGGCTTTAGTGATTTGTTTTTGTTCA...  54.590828   \n",
       "1      CACCGCCACGCCAGCAGCCAGACCGCCGCCGCCGACTGGCACAAAT...   0.748428   \n",
       "2      CGTTTGTCTGCGCTGTGTGCCGCAACGACCGCAGCAATGGGGGCCG...   0.706336   \n",
       "3      GCAGATACAACTCACACAATGCACCCGCTGTGTGAAATAAACAGAG...   1.061173   \n",
       "4      GTCAATATGCTCGTCAATCCATGCGATAAGATCCTGAATAATTTTC...   0.892963   \n",
       "...                                                  ...        ...   \n",
       "17762  AGAATCAGCAGTGTTAGGGCAGGCTGTCACCAATTTAATGCTTTCA...   1.332975   \n",
       "17763  AGACAAACAGTCTCAAGCACCCGTGGCTATTCTAGCTTAATAAGTT...   0.921965   \n",
       "17764  CTTAAGTAATCGAGAAAAAACAAATTTAATACAAAGGCTATTTGAT...   0.469512   \n",
       "17765  ACCGGTCAGCAAAATGGCGGTTACAGTTCGTGAAGCGGCGCTGGTG...   0.889601   \n",
       "17766  ACCGTTAACATTGCGCTCATTGACCATGGCAAACCGATTTTAGGCG...   0.000000   \n",
       "\n",
       "       RNA_exp_2  RNA_exp_sum_ave     DNA_sum  expn_med1  expn_med2  \\\n",
       "0      53.711399        54.151114   49.578588  55.212763  51.214980   \n",
       "1       0.669080         0.708754  113.905547   0.611601   0.666221   \n",
       "2       0.711005         0.708670   89.722359   0.639401   0.644214   \n",
       "3       0.855839         0.958506   54.485220   0.978561   0.885794   \n",
       "4       0.802813         0.847888   48.016955   0.778401   0.718905   \n",
       "...          ...              ...         ...        ...        ...   \n",
       "17762   1.417616         1.375296   23.106246   1.223201   1.279809   \n",
       "17763   0.881224         0.901595   28.526390   0.856241   0.924306   \n",
       "17764   0.601874         0.535693   10.462048   0.407734   0.369723   \n",
       "17765   1.078358         0.983979    0.934222   1.070301   0.693230   \n",
       "17766   0.924306         0.462153    0.506342   0.000000   0.924306   \n",
       "\n",
       "        expn_med  num_barcodes_mapped  num_barcodes_integrated     category  \\\n",
       "0      53.213872                   60                       47          tss   \n",
       "1       0.638911                  136                      106          tss   \n",
       "2       0.641807                  121                      100          tss   \n",
       "3       0.932177                   59                       52          tss   \n",
       "4       0.748653                   49                       41          tss   \n",
       "...          ...                  ...                      ...          ...   \n",
       "17762   1.251505                   41                       27  neg_control   \n",
       "17763   0.890274                   27                       23  neg_control   \n",
       "17764   0.388728                    8                        7  neg_control   \n",
       "17765   0.881765                    3                        3  neg_control   \n",
       "17766   0.462153                    4                        3  neg_control   \n",
       "\n",
       "       expn_med_fitted  expn_med_fitted_scaled    active  \n",
       "0            53.213872               60.237193    active  \n",
       "1             0.638911                0.723236  inactive  \n",
       "2             0.641807                0.726515  inactive  \n",
       "3             0.932177                1.055209    active  \n",
       "4             0.748653                0.847462  inactive  \n",
       "...                ...                     ...       ...  \n",
       "17762         1.251505                1.416682    active  \n",
       "17763         0.890274                1.007774    active  \n",
       "17764         0.388728                0.440034  inactive  \n",
       "17765         0.881765                0.998143  inactive  \n",
       "17766         0.462153                0.523150  inactive  \n",
       "\n",
       "[17767 rows x 20 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df7 = pd.read_csv(\"rLP5_Endo2_lb_expression_formatted_std.txt\", delimiter=\"\\t\")\n",
    "df7\n",
    "# Expression summary for the TSS promoter library (17,635 reported TSSs) in LB, integrated at the rLP5 site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "932b330d-c7e7-4308-9762-93fa7257ddb7",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m df8\u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrLP6_Endo2_lb_expression_formatted_std.txt\u001b[39m\u001b[38;5;124m\"\u001b[39m, delimiter\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m      2\u001b[0m df8\n",
      "\u001b[1;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "df8= pd.read_csv(\"rLP6_Endo2_lb_expression_formatted_std.txt\", delimiter=\"\\t\")\n",
    "df8\n",
    "# Expression summary for the TSS promoter library (17,635 reported TSSs) in LB, integrated at the rLP6 site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e622be4f-2e71-4cde-a574-3117d12a59d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1e5e855-2655-411c-bf11-1f2b6e4ba0f0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac1456fb-d46c-4565-95c9-9d00aa774365",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b64e09e-ca0d-4940-8ec0-9c1dae3537af",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
