{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6169c8c4-6c46-41c1-9c84-3bff9d2e2c8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import sbol2\n",
    "import os\n",
    "import requests\n",
    "from dotenv import load_dotenv\n",
    "import zipfile\n",
    "from __future__ import absolute_import, division, print_function\n",
    "import numpy as np\n",
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import tempfile\n",
    "from abc import abstractmethod, ABCMeta\n",
    "from sklearn.tree import DecisionTreeClassifier as scikit_DecisionTree\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.preprocessing import OneHotEncoder, StandardScaler\n",
    "from sklearn.model_selection import train_test_split, GridSearchCV\n",
    "import sys\n",
    "import argparse\n",
    "import time\n",
    "from sklearn.metrics import mean_squared_error, r2_score, accuracy_score\n",
    "# import xgboost as xgb\n",
    "from sklearn.metrics import mean_squared_error\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "current_dir = os.path.abspath('')\n",
    "data_path = os.path.join(current_dir, '..', 'data')\n",
    "attachments_path = os.path.join(current_dir, '..', 'attachments')\n",
    "pulled_attachments_path = os.path.join(current_dir, '..', 'pulled_attachments')\n",
    "sbol_path = os.path.join(current_dir, '..', 'sbol_data')\n",
    "downloaded_sbol_path = os.path.join(current_dir, '..', 'downloaded_sbol')\n",
    "original_data_path = os.path.join(data_path, 'original_data')\n",
    "scripts_path = os.path.join(current_dir, 'scripts')\n",
    "model_data_path = os.path.join(data_path, 'processed_data', 'replicated_models')\n",
    "model_output_path = os.path.join('..', 'model_outputs')\n",
    "\n",
    "load_dotenv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "05b881c7-b711-41ab-af99-5c3e0dcb7245",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(os.path.join(original_data_path, \"frag-rLP5_LB_expression.txt\"), delimiter=\" \")\n",
    "# Expression levels from genomic fragment MPRA (random 200–300 bp sheared fragments) in LB media\t\n",
    "df2 = pd.read_csv(os.path.join(original_data_path, \"frag-rLP5-M9_expression.txt\"), delimiter=\" \")\n",
    "# Expression levels from genomic fragment MPRA (random 200–300 bp sheared fragments) in M9 media at rLP5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a0741eed-bc04-4391-a3f1-cef1fd6aaaf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df3 = pd.read_csv(os.path.join(original_data_path, \"fLP3_Endo2_lb_expression_formatted.txt\"), delimiter=\" \")\n",
    "df3 = df3.dropna(subset=['start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b1e36732-0637-4755-b691-821f89a170ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>tss_name</th>\n",
       "      <th>tss_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>variant</th>\n",
       "      <th>RNA_exp_sum_1_1</th>\n",
       "      <th>RNA_exp_sum_1_2</th>\n",
       "      <th>RNA_exp_sum_2_1</th>\n",
       "      <th>RNA_exp_sum_2_2</th>\n",
       "      <th>RNA_exp_sum_1</th>\n",
       "      <th>RNA_exp_sum_2</th>\n",
       "      <th>RNA_exp_sum_ave</th>\n",
       "      <th>DNA_1</th>\n",
       "      <th>DNA_2</th>\n",
       "      <th>DNA_ave</th>\n",
       "      <th>expn_med</th>\n",
       "      <th>num_barcodes_mapped</th>\n",
       "      <th>num_barcodes_integrated</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSS_11125_storz_regulondb,2945404,+</td>\n",
       "      <td>TSS_11125_storz_regulondb</td>\n",
       "      <td>2945404.0</td>\n",
       "      <td>+</td>\n",
       "      <td>2945284.0</td>\n",
       "      <td>2945434.0</td>\n",
       "      <td>GCAAATTTTGCACAAAAAATAGGCTTTAGTGATTTGTTTTTGTTCA...</td>\n",
       "      <td>43.807236</td>\n",
       "      <td>55.833039</td>\n",
       "      <td>54.661823</td>\n",
       "      <td>49.577995</td>\n",
       "      <td>49.820137</td>\n",
       "      <td>52.119909</td>\n",
       "      <td>50.970023</td>\n",
       "      <td>31.625906</td>\n",
       "      <td>35.101752</td>\n",
       "      <td>33.363829</td>\n",
       "      <td>52.010547</td>\n",
       "      <td>50</td>\n",
       "      <td>28</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TSS_1218_storz,289782,+</td>\n",
       "      <td>TSS_1218_storz</td>\n",
       "      <td>289782.0</td>\n",
       "      <td>+</td>\n",
       "      <td>289662.0</td>\n",
       "      <td>289812.0</td>\n",
       "      <td>GTATCTGCCTCCGATTCTCTGCAGAAGCAGAAAGACATTGGATCGA...</td>\n",
       "      <td>0.562627</td>\n",
       "      <td>0.549348</td>\n",
       "      <td>0.816074</td>\n",
       "      <td>0.585183</td>\n",
       "      <td>0.555988</td>\n",
       "      <td>0.700629</td>\n",
       "      <td>0.628308</td>\n",
       "      <td>36.298833</td>\n",
       "      <td>31.606403</td>\n",
       "      <td>33.952618</td>\n",
       "      <td>0.565387</td>\n",
       "      <td>53</td>\n",
       "      <td>26</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TSS_12352_storz,3253602,-</td>\n",
       "      <td>TSS_12352_storz</td>\n",
       "      <td>3253602.0</td>\n",
       "      <td>-</td>\n",
       "      <td>3253572.0</td>\n",
       "      <td>3253722.0</td>\n",
       "      <td>CGTTTGTCTGCGCTGTGTGCCGCAACGACCGCAGCAATGGGGGCCG...</td>\n",
       "      <td>0.516503</td>\n",
       "      <td>0.717569</td>\n",
       "      <td>0.658121</td>\n",
       "      <td>0.587724</td>\n",
       "      <td>0.617036</td>\n",
       "      <td>0.622923</td>\n",
       "      <td>0.619979</td>\n",
       "      <td>72.105393</td>\n",
       "      <td>72.498105</td>\n",
       "      <td>72.301749</td>\n",
       "      <td>0.577755</td>\n",
       "      <td>108</td>\n",
       "      <td>54</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TSS_6478_storz,1684674,-</td>\n",
       "      <td>TSS_6478_storz</td>\n",
       "      <td>1684674.0</td>\n",
       "      <td>-</td>\n",
       "      <td>1684644.0</td>\n",
       "      <td>1684794.0</td>\n",
       "      <td>GCAGATACAACTCACACAATGCACCCGCTGTGTGAAATAAACAGAG...</td>\n",
       "      <td>1.006087</td>\n",
       "      <td>1.084363</td>\n",
       "      <td>0.763596</td>\n",
       "      <td>0.900281</td>\n",
       "      <td>1.045225</td>\n",
       "      <td>0.831939</td>\n",
       "      <td>0.938582</td>\n",
       "      <td>35.065895</td>\n",
       "      <td>37.408634</td>\n",
       "      <td>36.237265</td>\n",
       "      <td>0.936175</td>\n",
       "      <td>56</td>\n",
       "      <td>27</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TSS_2956_storz,770940,+</td>\n",
       "      <td>TSS_2956_storz</td>\n",
       "      <td>770940.0</td>\n",
       "      <td>+</td>\n",
       "      <td>770820.0</td>\n",
       "      <td>770970.0</td>\n",
       "      <td>TTATAAAGATATGACCAAGTTCTGGGGCAAGTTGTTTGGTATCAAC...</td>\n",
       "      <td>0.715541</td>\n",
       "      <td>0.744280</td>\n",
       "      <td>0.679240</td>\n",
       "      <td>0.634099</td>\n",
       "      <td>0.729911</td>\n",
       "      <td>0.656670</td>\n",
       "      <td>0.693290</td>\n",
       "      <td>79.950519</td>\n",
       "      <td>78.564136</td>\n",
       "      <td>79.257327</td>\n",
       "      <td>0.632548</td>\n",
       "      <td>116</td>\n",
       "      <td>59</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17533</th>\n",
       "      <td>TSS_6128_storz,1599413,+</td>\n",
       "      <td>TSS_6128_storz</td>\n",
       "      <td>1599413.0</td>\n",
       "      <td>+</td>\n",
       "      <td>1599293.0</td>\n",
       "      <td>1599443.0</td>\n",
       "      <td>AATCTTTAATGCAATTGTTCAGTTCTTGCTCATTTATATCTGTGAT...</td>\n",
       "      <td>0.339038</td>\n",
       "      <td>0.703175</td>\n",
       "      <td>0.504604</td>\n",
       "      <td>1.300406</td>\n",
       "      <td>0.521106</td>\n",
       "      <td>0.902505</td>\n",
       "      <td>0.711806</td>\n",
       "      <td>3.589814</td>\n",
       "      <td>3.511266</td>\n",
       "      <td>3.550540</td>\n",
       "      <td>0.574114</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17534</th>\n",
       "      <td>TSS_8744_storz,2304785,-</td>\n",
       "      <td>TSS_8744_storz</td>\n",
       "      <td>2304785.0</td>\n",
       "      <td>-</td>\n",
       "      <td>2304755.0</td>\n",
       "      <td>2304905.0</td>\n",
       "      <td>AAATTATAATGTTTCTAAAATTAGAATATAATTTATAAACATTATT...</td>\n",
       "      <td>1.112467</td>\n",
       "      <td>1.246183</td>\n",
       "      <td>1.027605</td>\n",
       "      <td>1.300406</td>\n",
       "      <td>1.179325</td>\n",
       "      <td>1.164006</td>\n",
       "      <td>1.171665</td>\n",
       "      <td>4.678904</td>\n",
       "      <td>4.427289</td>\n",
       "      <td>4.553096</td>\n",
       "      <td>1.234837</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17535</th>\n",
       "      <td>TSS_14532_regulondb,3740743,+</td>\n",
       "      <td>TSS_14532_regulondb</td>\n",
       "      <td>3740743.0</td>\n",
       "      <td>+</td>\n",
       "      <td>3740623.0</td>\n",
       "      <td>3740773.0</td>\n",
       "      <td>ATTGATTTTTATGTCTTTTGAAATTCATCAATCAGATTGCCTTGTT...</td>\n",
       "      <td>1.695188</td>\n",
       "      <td>1.104990</td>\n",
       "      <td>0.446785</td>\n",
       "      <td>1.253963</td>\n",
       "      <td>1.400089</td>\n",
       "      <td>0.850374</td>\n",
       "      <td>1.125232</td>\n",
       "      <td>2.379770</td>\n",
       "      <td>1.978044</td>\n",
       "      <td>2.178907</td>\n",
       "      <td>1.168341</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17536</th>\n",
       "      <td>TSS_5145_wanner,1306695,-</td>\n",
       "      <td>TSS_5145_wanner</td>\n",
       "      <td>1306695.0</td>\n",
       "      <td>-</td>\n",
       "      <td>1306665.0</td>\n",
       "      <td>1306815.0</td>\n",
       "      <td>TCATAGATTAGAAATCCTTTTGAAAGCGCATTGTTATGAGTTTACG...</td>\n",
       "      <td>1.331934</td>\n",
       "      <td>1.031324</td>\n",
       "      <td>1.572683</td>\n",
       "      <td>1.300406</td>\n",
       "      <td>1.181629</td>\n",
       "      <td>1.436545</td>\n",
       "      <td>1.309087</td>\n",
       "      <td>3.428542</td>\n",
       "      <td>2.855406</td>\n",
       "      <td>3.141974</td>\n",
       "      <td>1.409033</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17537</th>\n",
       "      <td>TSS_5146_regulondb,1306785,+</td>\n",
       "      <td>TSS_5146_regulondb</td>\n",
       "      <td>1306785.0</td>\n",
       "      <td>+</td>\n",
       "      <td>1306665.0</td>\n",
       "      <td>1306815.0</td>\n",
       "      <td>TCATAGATTAGAAATCCTTTTGAAAGCGCATTGTTATGAGTTTACG...</td>\n",
       "      <td>1.331934</td>\n",
       "      <td>1.031324</td>\n",
       "      <td>1.572683</td>\n",
       "      <td>1.300406</td>\n",
       "      <td>1.181629</td>\n",
       "      <td>1.436545</td>\n",
       "      <td>1.309087</td>\n",
       "      <td>3.428542</td>\n",
       "      <td>2.855406</td>\n",
       "      <td>3.141974</td>\n",
       "      <td>1.409033</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "      <td>tss</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17057 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      name                   tss_name  \\\n",
       "0      TSS_11125_storz_regulondb,2945404,+  TSS_11125_storz_regulondb   \n",
       "1                  TSS_1218_storz,289782,+             TSS_1218_storz   \n",
       "2                TSS_12352_storz,3253602,-            TSS_12352_storz   \n",
       "3                 TSS_6478_storz,1684674,-             TSS_6478_storz   \n",
       "4                  TSS_2956_storz,770940,+             TSS_2956_storz   \n",
       "...                                    ...                        ...   \n",
       "17533             TSS_6128_storz,1599413,+             TSS_6128_storz   \n",
       "17534             TSS_8744_storz,2304785,-             TSS_8744_storz   \n",
       "17535        TSS_14532_regulondb,3740743,+        TSS_14532_regulondb   \n",
       "17536            TSS_5145_wanner,1306695,-            TSS_5145_wanner   \n",
       "17537         TSS_5146_regulondb,1306785,+         TSS_5146_regulondb   \n",
       "\n",
       "       tss_position strand      start        end  \\\n",
       "0         2945404.0      +  2945284.0  2945434.0   \n",
       "1          289782.0      +   289662.0   289812.0   \n",
       "2         3253602.0      -  3253572.0  3253722.0   \n",
       "3         1684674.0      -  1684644.0  1684794.0   \n",
       "4          770940.0      +   770820.0   770970.0   \n",
       "...             ...    ...        ...        ...   \n",
       "17533     1599413.0      +  1599293.0  1599443.0   \n",
       "17534     2304785.0      -  2304755.0  2304905.0   \n",
       "17535     3740743.0      +  3740623.0  3740773.0   \n",
       "17536     1306695.0      -  1306665.0  1306815.0   \n",
       "17537     1306785.0      +  1306665.0  1306815.0   \n",
       "\n",
       "                                                 variant  RNA_exp_sum_1_1  \\\n",
       "0      GCAAATTTTGCACAAAAAATAGGCTTTAGTGATTTGTTTTTGTTCA...        43.807236   \n",
       "1      GTATCTGCCTCCGATTCTCTGCAGAAGCAGAAAGACATTGGATCGA...         0.562627   \n",
       "2      CGTTTGTCTGCGCTGTGTGCCGCAACGACCGCAGCAATGGGGGCCG...         0.516503   \n",
       "3      GCAGATACAACTCACACAATGCACCCGCTGTGTGAAATAAACAGAG...         1.006087   \n",
       "4      TTATAAAGATATGACCAAGTTCTGGGGCAAGTTGTTTGGTATCAAC...         0.715541   \n",
       "...                                                  ...              ...   \n",
       "17533  AATCTTTAATGCAATTGTTCAGTTCTTGCTCATTTATATCTGTGAT...         0.339038   \n",
       "17534  AAATTATAATGTTTCTAAAATTAGAATATAATTTATAAACATTATT...         1.112467   \n",
       "17535  ATTGATTTTTATGTCTTTTGAAATTCATCAATCAGATTGCCTTGTT...         1.695188   \n",
       "17536  TCATAGATTAGAAATCCTTTTGAAAGCGCATTGTTATGAGTTTACG...         1.331934   \n",
       "17537  TCATAGATTAGAAATCCTTTTGAAAGCGCATTGTTATGAGTTTACG...         1.331934   \n",
       "\n",
       "       RNA_exp_sum_1_2  RNA_exp_sum_2_1  RNA_exp_sum_2_2  RNA_exp_sum_1  \\\n",
       "0            55.833039        54.661823        49.577995      49.820137   \n",
       "1             0.549348         0.816074         0.585183       0.555988   \n",
       "2             0.717569         0.658121         0.587724       0.617036   \n",
       "3             1.084363         0.763596         0.900281       1.045225   \n",
       "4             0.744280         0.679240         0.634099       0.729911   \n",
       "...                ...              ...              ...            ...   \n",
       "17533         0.703175         0.504604         1.300406       0.521106   \n",
       "17534         1.246183         1.027605         1.300406       1.179325   \n",
       "17535         1.104990         0.446785         1.253963       1.400089   \n",
       "17536         1.031324         1.572683         1.300406       1.181629   \n",
       "17537         1.031324         1.572683         1.300406       1.181629   \n",
       "\n",
       "       RNA_exp_sum_2  RNA_exp_sum_ave      DNA_1      DNA_2    DNA_ave  \\\n",
       "0          52.119909        50.970023  31.625906  35.101752  33.363829   \n",
       "1           0.700629         0.628308  36.298833  31.606403  33.952618   \n",
       "2           0.622923         0.619979  72.105393  72.498105  72.301749   \n",
       "3           0.831939         0.938582  35.065895  37.408634  36.237265   \n",
       "4           0.656670         0.693290  79.950519  78.564136  79.257327   \n",
       "...              ...              ...        ...        ...        ...   \n",
       "17533       0.902505         0.711806   3.589814   3.511266   3.550540   \n",
       "17534       1.164006         1.171665   4.678904   4.427289   4.553096   \n",
       "17535       0.850374         1.125232   2.379770   1.978044   2.178907   \n",
       "17536       1.436545         1.309087   3.428542   2.855406   3.141974   \n",
       "17537       1.436545         1.309087   3.428542   2.855406   3.141974   \n",
       "\n",
       "        expn_med  num_barcodes_mapped  num_barcodes_integrated category  \n",
       "0      52.010547                   50                       28      tss  \n",
       "1       0.565387                   53                       26      tss  \n",
       "2       0.577755                  108                       54      tss  \n",
       "3       0.936175                   56                       27      tss  \n",
       "4       0.632548                  116                       59      tss  \n",
       "...          ...                  ...                      ...      ...  \n",
       "17533   0.574114                    9                        3      tss  \n",
       "17534   1.234837                    3                        3      tss  \n",
       "17535   1.168341                    3                        3      tss  \n",
       "17536   1.409033                   14                        4      tss  \n",
       "17537   1.409033                   14                        4      tss  \n",
       "\n",
       "[17057 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "761f55c1-a68b-485f-826f-09a11af50b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_sbol_attempt2_doc(df, file_path, media=\"LB\"):\n",
    "    doc = sbol2.Document()\n",
    "    sbol2.setHomespace('http://github.com/cywlol/promoters')\n",
    "    doc.displayId = \"E_coli_promoters\"\n",
    "\n",
    "    media_label_MD = media\n",
    "    chassis_label_MD = \"E_coli_chassis\"\n",
    "    sample_design_labels = []\n",
    "    \n",
    "    # define the media once, then have it as a reference for each promoter row\n",
    "    media_md = sbol2.ModuleDefinition(media_label_MD)\n",
    "    media_md.addRole(\"http://identifiers.org/ncit/NCIT:C48164\")\n",
    "    media_md.version = \"1\"\n",
    "    doc.addModuleDefinition(media_md)\n",
    "\n",
    "    # define the chassis once, then have it as a reference for each promoter row\n",
    "    chassis_md = sbol2.ModuleDefinition(chassis_label_MD)\n",
    "    chassis_md.addRole(\"http://identifiers.org/ncit/NCIT:C14419\")\n",
    "    chassis_md.wasDerivedFrom = [\"https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=511145\"]\n",
    "    chassis_md.version = \"1\"\n",
    "    doc.addModuleDefinition(chassis_md)\n",
    "\n",
    "    for i, row in df.iterrows():\n",
    "        fragment_seq = row[\"variant\"]\n",
    "        start = row[\"start\"]\n",
    "        end = row[\"end\"]\n",
    "        strand = row[\"strand\"]\n",
    "        RNA_exp_ave = row[\"expn_med\"]\n",
    "\n",
    "        # Establish the identities of each component, ensuring that each one is unique\n",
    "        promoter_label_CD = f\"promoter_definition_{i}\"\n",
    "        promoter_label_FC = f\"promoter_fc_{i}\"\n",
    "        promoter_seq_label = f\"promoter_seq_{i}\"\n",
    "        sample_design_label_MD = f\"sample_design_definition_{i}\"\n",
    "        chassis_module_label = f\"chassis_module_{i}\"\n",
    "        circuit_label_CD = f\"circuit_definition_{i}\"\n",
    "        strain_label_MD = f\"strain_definition_{i}\"\n",
    "        circuit_label_FC = f\"cicuit_fc_{i}\"\n",
    "        measurement_label = f\"exp_ave_{i}\"\n",
    "        interaction_label = f\"interaction_{i}\"\n",
    "        gfp_fc_label = f\"gfp_{i}\"\n",
    "        promoter_location_label = f\"promoter_{i}_location\"\n",
    "        strain_module_label = f\"strain_module_{i}\"\n",
    "        media_module_label = f\"media_module_{i}\"\n",
    "        # define gfp once\n",
    "        gfp_component_label = f\"gfp_{i}\"\n",
    "\n",
    "        sample_design_labels.append(sample_design_label_MD)\n",
    "        #engr_region_label_CD = f\"engr_region_{i}\"\n",
    "        #location_promoter_annotation = f\"location_promoter_annotation_{i}\"\n",
    "        #location_promoter_label = f\"location_promoter_label_{i}\"\n",
    "\n",
    "\n",
    "        '''\n",
    "        # Engineered Region\n",
    "        engineered_cd = sbol2.ComponentDefinition(engr_region_label_CD, sbol2.BIOPAX_DNA)\n",
    "        engineered_cd.roles = [\"https://identifiers.org/so/SO:0000804\"]\n",
    "        sub = engineered_cd.components.create('promoter')\n",
    "        sub.definition = promoter_cd.persistentIdentity\n",
    "        doc.addComponentDefinition(engineered_cd)\n",
    "        # No sequence?\n",
    "        '''\n",
    "\n",
    "\n",
    "        # Add strain module definition\n",
    "        strain_md = sbol2.ModuleDefinition(strain_label_MD)\n",
    "        strain_md.addRole(\"http://identifiers.org/ncit/NCIT:C14419\")\n",
    "        strain_md.version = \"1\"\n",
    "        doc.addModuleDefinition(strain_md)\n",
    "\n",
    "        chassis_module = sbol2.Module(chassis_module_label)\n",
    "        chassis_module.definition = chassis_md.identity\n",
    "        strain_md.modules.add(chassis_module)\n",
    "\n",
    "        # Sample Design\n",
    "        sample_md = sbol2.ModuleDefinition(sample_design_label_MD)\n",
    "        sample_md.addRole(\"http://identifiers.org/obo/OBI:0000073\")\n",
    "        sample_md.version = \"1\"\n",
    "        doc.addModuleDefinition(sample_md)\n",
    "\n",
    "        # Add strain module to sample design\n",
    "        strain_module = sbol2.Module(strain_module_label)\n",
    "        strain_module.definition = strain_md.identity\n",
    "        # Add media module to sample design\n",
    "        media_module = sbol2.Module(media_module_label)\n",
    "        media_module.definition = media_md.identity\n",
    "\n",
    "\n",
    "        sample_md.modules.add(media_module)\n",
    "        sample_md.modules.add(strain_module)\n",
    "\n",
    "        # Promoter and Sequence\n",
    "        promoter_cd = sbol2.ComponentDefinition(promoter_label_CD, sbol2.BIOPAX_DNA)\n",
    "        promoter_cd.roles = [sbol2.SO_PROMOTER]\n",
    "        promoter_cd.version = \"1\"\n",
    "        doc.addComponentDefinition(promoter_cd)\n",
    "        seq = sbol2.Sequence(promoter_seq_label, fragment_seq, sbol2.SBOL_ENCODING_IUPAC)\n",
    "        seq.version = \"1\"\n",
    "\n",
    "        doc.addSequence(seq)\n",
    "        promoter_cd.sequences = seq\n",
    "\n",
    "        # Circuit\n",
    "        circuit = sbol2.ComponentDefinition(circuit_label_CD, sbol2.BIOPAX_DNA)\n",
    "        circuit.roles = [\"https://identifiers.org/so/SO:0000804\"]\n",
    "        circuit.version = \"1\"\n",
    "\n",
    "        annotation = sbol2.SequenceAnnotation(promoter_location_label)\n",
    "        annotation.version = \"1\"\n",
    "        rng = sbol2.Range(\"prange\", start, end)\n",
    "        rng.version = \"1\"\n",
    "\n",
    "        if (strand == \"-\"):\n",
    "            rng.orientation = sbol2.SBOL_ORIENTATION_REVERSE_COMPLEMENT\n",
    "\n",
    "        annotation.locations.add(rng)\n",
    "        circuit.sequenceAnnotations.add(annotation)\n",
    "        promoter_comp = circuit.components.create('promoter')\n",
    "        promoter_comp.definition = promoter_cd.identity\n",
    "        doc.addComponentDefinition(circuit)\n",
    "\n",
    "\n",
    "        circuit_fc = sbol2.FunctionalComponent(circuit_label_FC)\n",
    "        circuit_fc.version = \"1\"\n",
    "        circuit_fc.definition = circuit.identity\n",
    "        circuit_fc.direction = sbol2.SBOL_DIRECTION_OUT\n",
    "        circuit_fc.access = sbol2.SBOL_ACCESS_PUBLIC\n",
    "        strain_md.functionalComponents.add(circuit_fc)\n",
    "\n",
    "        measurement3 = sbol2.Measurement(measurement_label, value=RNA_exp_ave, unit=\"http://www.ontology-of-units-of-measure.org/resource/om-2/RatioUnit\")\n",
    "        measurement3.version = \"1\"\n",
    "\n",
    "        gfp_protein = sbol2.ComponentDefinition(gfp_component_label, sbol2.BIOPAX_PROTEIN)\n",
    "        gfp_protein.roles=[sbol2.SO_CDS]\n",
    "        gfp_protein.version = \"1\"\n",
    "        doc.addComponentDefinition(gfp_protein)\n",
    "\n",
    "        gfp_fc = sbol2.FunctionalComponent(gfp_fc_label)\n",
    "        gfp_fc.version = \"1\"\n",
    "        gfp_fc.definition = gfp_protein.identity\n",
    "        gfp_fc.direction = sbol2.SBOL_DIRECTION_OUT\n",
    "        gfp_fc.access = sbol2.SBOL_ACCESS_PUBLIC\n",
    "\n",
    "\n",
    "        doc.add(measurement3)\n",
    "        m = gfp_fc.measurements.add(measurement3)\n",
    "\n",
    "        sample_md.functionalComponents.add(gfp_fc)\n",
    "        # exp_attachment = sbol2.Attachment(exp_label)\n",
    "        # exp_attachment.name = f'fragmentation expression at rLP5 for sample {i}'\n",
    "        # exp_attachment.description = 'CSV including the gene expression of the sequence: RNA_exp1, RNA_exp2, and its average.'\n",
    "        # exp_attachment.source = 'CSV_LINK_HERE' # update when added attachment to SBOL collection\n",
    "        # exp_attachment.format = 'https://identifiers.org/edam/format_3752'\n",
    "        # doc.addAttachment(exp_attachment)\n",
    "\n",
    "\n",
    "    report = doc.validate()\n",
    "    if (report == 'Valid.'):\n",
    "        doc.write(file_path)\n",
    "    else:\n",
    "        print(report)\n",
    "    return doc, chassis_label_MD, sample_design_labels\n",
    "\n",
    "\n",
    "        \n",
    "def partshop_attach_exp_data(synbio_username, collection_name, file_names, email, password, exp_data_labels, version=1):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    print(shop.login(email, password))\n",
    "    exp_labels = [\"ExperimentalData_\" + label for label in exp_data_labels]  \n",
    "    for label, file_name in zip(exp_labels, file_names):\n",
    "        path = os.path.join(attachments_path, file_name)\n",
    "        attachment_uri = f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{label}/{version}\"\n",
    "        shop.attachFile(attachment_uri, path)\n",
    "\n",
    "def partshop_attach_genome_to_md(synbio_username, collection_name, file_path, email, password, chassis_label, version=1):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    print(shop.login(email, password))\n",
    "\n",
    "    label = \"ModuleDefinition_\" + chassis_label\n",
    "    attachment_uri = f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{label}/{version}\"\n",
    "\n",
    "    shop.attachFile(attachment_uri, file_path)\n",
    "\n",
    "def create_synbio_collection(email, password, file_path, id, name, description, version='1'):\n",
    "    response = requests.post(\n",
    "        \"https://synbiohub.org/login\",\n",
    "        headers={\"Accept\": \"text/plain\"},\n",
    "        data={\"email\": email, \"password\": password}\n",
    "    )\n",
    "        \n",
    "    if response.ok:\n",
    "        token = response.text.strip() # theres a whitespace before the token for some reason\n",
    "        response = requests.post(\n",
    "        'https://synbiohub.org/submit',\n",
    "        headers={\n",
    "            'X-authorization': token,\n",
    "            'Accept': 'text/plain'\n",
    "        },\n",
    "        files={\n",
    "        'files': open(file_path,'rb'),\n",
    "        },\n",
    "        data={\n",
    "            'id': id,\n",
    "            'version' : version,\n",
    "            'name' :  name,\n",
    "            'description' : description,\n",
    "            'citations' : '',\n",
    "            'overwrite_merge' : '0'\n",
    "        },\n",
    "    \n",
    "    )\n",
    "    else:\n",
    "        print(\"Login failed:\", response.status_code)\n",
    "        print(response.text)\n",
    "\n",
    "def partshop_pull_collection(email, password, synbio_username, collection_name, file_path, version=1):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    doc = sbol2.Document()\n",
    "    shop.login(email, password)\n",
    "\n",
    "    collection_uri = f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{collection_name}_collection/{version}\"\n",
    "    s = shop.pull(collection_uri, doc)\n",
    "    \n",
    "    for obj in doc:\n",
    "        print(obj)   \n",
    "    \n",
    "    doc.write(file_path)\n",
    "    return doc\n",
    "\n",
    "def partshop_pull_sample_designs(email, password, synbio_username, collection_name, sample_design_labels, output_dir, prefix, version=1):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    shop.login(email, password)\n",
    "\n",
    "    for sample_design_label in sample_design_labels: \n",
    "        doc = sbol2.Document()\n",
    "     #f\"https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_0/1\n",
    "        collection_uri = f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{prefix}_{sample_design_label}/{version}\"\n",
    "        s = shop.pull(collection_uri, doc)\n",
    "        doc.write(os.path.join(output_dir, sample_design_label + \".xml\"))\n",
    "\n",
    "def new_partshop_pull_sample_designs(email, password, synbio_username, collection_name, sample_design_labels, output_dir, prefix, version=1):\n",
    "\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    doc = sbol2.Document()\n",
    "    shop.login(email, password)\n",
    "    uris = []\n",
    "\n",
    "    for sample_design_label in sample_design_labels: \n",
    "        uris.append(f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{prefix}_{sample_design_label}/{version}\")\n",
    "        \n",
    "    shop.pull(uris, doc)\n",
    "    doc.write(os.path.join(output_dir, sample_design_label + \".xml\"))\n",
    "        \n",
    "def partshop_submit(email, password, synbio_username, collection_name, file_path, version=1):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    doc = sbol2.Document()\n",
    "    shop.login(email, password)\n",
    "\n",
    "    collection_uri = f\"https://synbiohub.org/user/{synbio_username}/{collection_name}/{collection_name}_collection/{version}\"\n",
    "    s = shop.submit(collection_uri, doc)\n",
    "    \n",
    "    for obj in doc:\n",
    "        print(obj)   \n",
    "    \n",
    "    doc.write(file_path)\n",
    "\n",
    "def download_all_attachments(email, password, doc, file_path):\n",
    "    shop = sbol2.PartShop(\"https://synbiohub.org\")\n",
    "    shop.login(email, password)\n",
    "    for attachment in doc.attachments:\n",
    "        shop.downloadAttachment(attachment.identity, filepath=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6bf198d9-d05a-4a6a-9d2e-fad529266188",
   "metadata": {},
   "outputs": [],
   "source": [
    "ecoli_genome_file_name = \"E. coli.fasta\"\n",
    "env_email = os.getenv(\"SYNBIO_EMAIL\")\n",
    "env_password = os.getenv(\"SYNBIO_PASSWORD\")\n",
    "\n",
    "username = \"cywong\"\n",
    "output_name = 'ecolipromoter_one.xml'\n",
    "id = \"Ecolipromoterexpdataalt\"\n",
    "name = \"E coli promoter data exploration alternative\"\n",
    "description = \"A collection containing the extracted E coli data from paper\"\n",
    "sbol_file_name = output_name\n",
    "imported_sbol_file_name = \"promoters_import.xml\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0de59a6b-c443-4788-a8f0-0d1235fab7e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "doc, chassis_label, sample_design_labels = create_sbol_attempt2_doc(df3, os.path.join(sbol_path, output_name))       \n",
    "\n",
    "# # # #partshop_attach_exp_data(username, id, attachment_file_names, env_email, env_password, exp_labels)\n",
    "# partshop_attach_genome_to_md(username, id, os.path.join(attachments_path, ecoli_genome_file_name), env_email, env_password, chassis_label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b2dd1077-8609-4a9d-acee-059b0778f579",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_synbio_collection(env_email, env_password, os.path.join(sbol_path, sbol_file_name), id, name, description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b9268715-c11e-4fe6-adfe-50159bb4772a",
   "metadata": {},
   "outputs": [],
   "source": [
    "partshop_pull_sample_designs(env_email, env_password, username, id, sample_design_labels, sbol_path, \"ModuleDefinition\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36c0cfd1-3334-4808-bb9e-2b20ef3c35b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "doc = partshop_pull_collection(env_email, env_password, username, id, os.path.join(downloaded_sbol_path, imported_sbol_file_name), version=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6bea158-be70-4910-9e4d-2978b24bab05",
   "metadata": {},
   "outputs": [],
   "source": [
    "doc2 = sbol2.Document()\n",
    "doc2.addModuleDefinition(doc.get(\"https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_0/1\"))\n",
    "doc2.write(\"dd.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b68c3428-b161-4c9a-868c-c671fede0029",
   "metadata": {},
   "outputs": [],
   "source": [
    "sd = doc.get(\"https://synbiohub.org/user/cywong/Ecolipromoterexpdataalt/ModuleDefinition_sample_design_definition_0/1\")\n",
    "doc = sbol2.Document()\n",
    "doc.addModuleDefinition(sd)\n",
    "doc.write(os.path.join(sbol_path, \"sample_design.xml\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7df4bf5-981f-44ff-8a1a-573dd9996bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# download_all_attachments(env_email, env_password, doc, pulled_attachments_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48b8bd54-df14-4eba-a637-138bb295fcb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_df_from_sbol_doc(doc, pulled_attachments_path):\n",
    "    df_new = pd.DataFrame()\n",
    "    for exp in doc.experimentalData:\n",
    "        attachment = doc.get(exp.attachments[0])\n",
    "        sample_design = doc.get(exp.wasDerivedFrom[0])\n",
    "    \n",
    "        # may need to be changed\n",
    "        strain = doc.get(sample_design).modules[0] if 'strain' in doc.get(sample_design).modules[0].identity else doc.get(sample_design).modules[1]\n",
    "        promoter =  doc.get(doc.get(strain.definition).functionalComponents[0].definition)\n",
    "        promoter_seq = doc.get(promoter.sequences[0]).elements\n",
    "        \n",
    "        df1 = pd.read_csv(os.path.join(pulled_attachments_path, attachment.name))\n",
    "        df1['Sequence'] = promoter_seq\n",
    "        df_new = pd.concat([df_new, df1], ignore_index=True)\n",
    "\n",
    "    return df_new\n",
    "\n",
    "\n",
    "# for mod in doc.moduleDefinitions:\n",
    "#     print(\"Mod: \", mod.identity)\n",
    "# for attachment in doc.attachments:\n",
    "#     print(\"Attachment: \", attachment.source) \n",
    "    \n",
    "# for seq in doc.sequences:\n",
    "#     print(\"Seq:\", seq.elements) \n",
    "\n",
    "doc = sbol2.Document()\n",
    "doc.read(os.path.join(downloaded_sbol_path, imported_sbol_file_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e292392c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Model(object):\n",
    "    __metaclass__ = ABCMeta\n",
    "\n",
    "    @abstractmethod\n",
    "    def __init__(self, **hyperparameters):\n",
    "        pass\n",
    "\n",
    "    @abstractmethod\n",
    "    def train(self, X, y, validation_data):\n",
    "        pass\n",
    "\n",
    "    @abstractmethod\n",
    "    def predict(self, X):\n",
    "        pass\n",
    "\n",
    "    def test(self, X, y):\n",
    "        return self.evaluate(X, y)\n",
    "        # return ClassificationResult(y, self.predict(X))\n",
    "\n",
    "    def score(self, X, y):\n",
    "        pass\n",
    "\n",
    "class DecisionTree(Model):\n",
    "\n",
    "    def __init__(self):\n",
    "        self.classifier = scikit_DecisionTree()\n",
    "\n",
    "    def train(self, X, y, validation_data=None):\n",
    "        self.classifier.fit(X, y)\n",
    "\n",
    "    def predict(self, X):\n",
    "        predictions = np.asarray(self.classifier.predict_proba(X))[..., 1]\n",
    "        if len(predictions.shape) == 2:  # multitask\n",
    "            predictions = predictions.T\n",
    "        else:  # single-task\n",
    "            predictions = np.expand_dims(predictions, 1)\n",
    "        return predictions\n",
    "    \n",
    "\n",
    "class RandomForestRegression(DecisionTree):\n",
    "    def __init__(self):\n",
    "        self.regressor = RandomForestRegressor(n_estimators=100)\n",
    "\n",
    "    def train(self, X, y, validation_data=None):\n",
    "        # X shape: n_samples, n_features\n",
    "        # y shape: n_samples\n",
    "        self.regressor.fit(X, y)\n",
    "\n",
    "    def predict(self, X):\n",
    "        return self.regressor.predict(X)\n",
    "\n",
    "    def score(self, X, y):\n",
    "        predictions = np.squeeze(self.regressor.predict(X))\n",
    "        return np.corrcoef(predictions, y)[0,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c0490a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_df = pd.read_csv(os.path.join(model_data_path, \"tss_expression_model_format_train_genome_split.txt\"), sep='\\t', header=None, names=[\"Sequence\", \"Expression\"])\n",
    "test_df = pd.read_csv(os.path.join(model_data_path, \"tss_expression_model_format_test_genome_split.txt\"), sep='\\t',  header=None, names=[\"Sequence\", \"Expression\"])\n",
    "\n",
    "X_train, y_train = process_seqs(train_df, 150, \"Sequence\", \"Expression\")\n",
    "X_test, y_test = process_seqs(test_df, 150, \"Sequence\", \"Expression\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9826fbb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_RandomForest(X_train, X_test, y_train, y_test):\n",
    "    print(\"Running random forest regression...\")\n",
    "    model = RandomForestRegression()\n",
    "    model.train(X_train, y_train)\n",
    "    predictions = model.predict(X_test)\n",
    "    \n",
    "    with open(\"outputs.txt\", 'w') as outfile:\n",
    "        for i in range(len(predictions)):\n",
    "            outfile.write(str(float(predictions[i])) + '\\t' +\n",
    "                      str(float(y_test[i])) + '\\n')\n",
    "    \n",
    "    score = model.score(X_test, y_test)\n",
    "    print(\"Score:\", score)\n",
    "    return model, predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb56143f",
   "metadata": {},
   "outputs": [],
   "source": [
    "model, predictions = train_RandomForest(X_train, X_test, y_train, y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ab152ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "mse = mean_squared_error(y_test, predictions)\n",
    "r_squared = r2_score(y_test, predictions)\n",
    "rmse = np.sqrt(mse)\n",
    "\n",
    "print(f\"Mean Squared Error (MSE): {mse}\")\n",
    "print(f\"Root Mean Square Error (RMSE): {rmse}\")\n",
    "print(f\"R^2: {r_squared}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11b575e9-ebaf-4789-9bc9-9b67f80b4659",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.iloc[0:5000].sample(frac=1, random_state=42).reset_index(drop=True)\n",
    "df['strand'] = df['strand'].map({'+': 0, '-': 1})\n",
    "df.drop(['start', 'end'], axis=1, inplace=True)\n",
    "\n",
    "\n",
    "columns_to_scale = ['RNA_exp_1', 'DNA_sum_1', 'DNA_sum_2', 'DNA_ave', 'variation']\n",
    "scaler = StandardScaler()\n",
    "df[columns_to_scale] = scaler.fit_transform(df[columns_to_scale])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "143f885b-13a5-4b8c-a263-f086a36bd56e",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_sequences = df3['variant']\n",
    "y = df3['expn_med'].values.reshape(-1, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a94af75e-ea56-458b-a2ad-703fb4e1d37e",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_sequences_ohe = []\n",
    "for i in range(len(X_sequences)):\n",
    "    ohe = one_hot_encode_modern(X_sequences.iloc[i])\n",
    "    X_sequences_ohe.append(ohe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "db2f48ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(X_sequences_ohe)\n",
    "y_scaled = StandardScaler().fit_transform(y)\n",
    "y_scaled = y_scaled.ravel()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6368cbb7-615d-4ae1-831d-e00891f1b7bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = []\n",
    "for i in range(len(df)):\n",
    "    seq = pad_sequence(df.loc[i, 'fragment'], 200)\n",
    "    ohe = one_hot_encode_modern(seq)\n",
    "    a.append(ohe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d328caa-daac-450d-91d9-1ef63b3f3867",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = np.array(a)\n",
    "X1 = X.reshape(X.shape[0], -1)\n",
    "X2 = df.drop(['RNA_exp_ave', 'fragment'], axis=1).values\n",
    "X = np.concatenate([X1, X2], axis=1)\n",
    "y = np.log1p(df['RNA_exp_ave'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "738c09ab-7b61-4b6e-99e8-81b41c9c9b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    " df['RNA_exp_ave'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88a9f35d-a919-4fd8-b1ff-8e169f10ec69",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "xgb_model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25c4f162-f50b-4994-b0a9-abacb4170c79",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "226ea739-e41a-46b1-98b8-cad2a3c3088d",
   "metadata": {},
   "outputs": [],
   "source": [
    "param_grid = {\n",
    "    'n_estimators': [50, 100],\n",
    "    'max_depth': [3, 7],\n",
    "    'learning_rate': [0.01, 0.1],\n",
    "    'colsample_bytree': [0.8, 1.0]\n",
    "}\n",
    "\n",
    "grid_search = GridSearchCV(\n",
    "    estimator=xgb_model,\n",
    "    param_grid=param_grid,\n",
    "    scoring='neg_mean_squared_error',  \n",
    "    cv=3,\n",
    "    verbose=1,\n",
    ")\n",
    "\n",
    "grid = grid_search.fit(X_train, y_train)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1f48c49-1a43-41d1-879d-c381cf4ee734",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Best parameters:\", grid_search.best_params_)\n",
    "\n",
    "best_model = grid_search.best_estimator_\n",
    "y_pred = best_model.predict(X_test)\n",
    "\n",
    "mse = mean_squared_error(y_test, y_pred)\n",
    "r_sq = r2_score(y_test, y_pred)\n",
    "print(\"Test MSE:\", mse)\n",
    "print(\"R sq:\", r_sq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "002109e6-091b-41ee-ac79-a454a9d004f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "xgb_model = xgb.XGBRegressor(objective='reg:squarederror', random_state=42, colsample_bytree=1, learning_rate=0.1, max_depth=3, n_estimators=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5d6c413-fa53-4b38-8139-8f700e9a36b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "xgb_model.fit(X_train, y_train)\n",
    "\n",
    "preds = xgb_model.predict(X_test)\n",
    "\n",
    "r_sq = r2_score(y_test, preds) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59dad312-a450-4c42-abcb-20f5887139fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_sq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "3ac69642",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(X_sequences_ohe, y_scaled, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "e64bb0df",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train = np.array(X_train) \n",
    "X_test = np.array(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "bd2022db",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train = X_train.reshape(X_train.shape[0], -1)\n",
    "X_test = X_test.reshape(X_test.shape[0], -1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f7ba8398",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_train = y_train.ravel()\n",
    "y_test = y_test.ravel()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "6381b637",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-0.09600204,  0.05270001, -0.07746396, ..., -0.10162205,\n",
       "       -0.09888044, -0.10658665])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f17db0ed-259e-4f14-bbc8-15d7c79d8728",
   "metadata": {},
   "outputs": [],
   "source": [
    "rfc = RandomForestRegression()\n",
    "rfc.train(X_train, y_train)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "05659b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "preds = rfc.predict(X_test)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "440bc2fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_sq = r2_score(y_test, preds) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "4abbe0da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7960463681705944"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_squared_error(y_test, preds)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "f2b7107c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.20183621111801853"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r2_score(y_test, preds) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8902c1f5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
