{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6169c8c4-6c46-41c1-9c84-3bff9d2e2c8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sbol2\n",
    "import os\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f299421b-7210-4145-94fa-c35d015df645",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_dir = os.path.abspath('')\n",
    "data_path = os.path.join(current_dir, '..', 'data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "05b881c7-b711-41ab-af99-5c3e0dcb7245",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fragment</th>\n",
       "      <th>RNA_exp_1</th>\n",
       "      <th>RNA_exp_2</th>\n",
       "      <th>RNA_exp_ave</th>\n",
       "      <th>DNA_sum_1</th>\n",
       "      <th>DNA_sum_2</th>\n",
       "      <th>DNA_ave</th>\n",
       "      <th>num_mapped_barcodes</th>\n",
       "      <th>num_integrated_barcodes</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTTCCAACCCATGGCGCGTGCGTACATAAAGGTTTCGGACGCGCGC...</td>\n",
       "      <td>1.316755</td>\n",
       "      <td>1.635498</td>\n",
       "      <td>1.476127</td>\n",
       "      <td>1.006226</td>\n",
       "      <td>1.006226</td>\n",
       "      <td>1.006226</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1838405</td>\n",
       "      <td>1838702</td>\n",
       "      <td>-</td>\n",
       "      <td>0.312743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGACTTCTTCGTGAACTTGCTGGATATGCGTTACGAGTGGAAAGCG...</td>\n",
       "      <td>0.776481</td>\n",
       "      <td>0.881533</td>\n",
       "      <td>0.829007</td>\n",
       "      <td>2.227730</td>\n",
       "      <td>2.227730</td>\n",
       "      <td>2.227730</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4133776</td>\n",
       "      <td>4134009</td>\n",
       "      <td>+</td>\n",
       "      <td>0.183064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TGGGGGGATGTATGGGTACGTTGTAATTAGGGATTTAACGAATTAG...</td>\n",
       "      <td>0.530471</td>\n",
       "      <td>1.426090</td>\n",
       "      <td>0.978280</td>\n",
       "      <td>2.721586</td>\n",
       "      <td>2.721586</td>\n",
       "      <td>2.721586</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2862818</td>\n",
       "      <td>2863076</td>\n",
       "      <td>-</td>\n",
       "      <td>1.426718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAATAAAATTCAGCAATCAATTAATGCCTTACATCAACATGGCA...</td>\n",
       "      <td>2.392780</td>\n",
       "      <td>2.118698</td>\n",
       "      <td>2.255739</td>\n",
       "      <td>5.572886</td>\n",
       "      <td>5.572886</td>\n",
       "      <td>5.572886</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3798297</td>\n",
       "      <td>3798568</td>\n",
       "      <td>-</td>\n",
       "      <td>0.175510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTTTTTGTCGTTGACCTCACCATGTCGATCACTGTGCCTGTATCCC...</td>\n",
       "      <td>0.664520</td>\n",
       "      <td>0.619367</td>\n",
       "      <td>0.641943</td>\n",
       "      <td>1.163899</td>\n",
       "      <td>1.163899</td>\n",
       "      <td>1.163899</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3657099</td>\n",
       "      <td>3657389</td>\n",
       "      <td>+</td>\n",
       "      <td>0.101518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321118</th>\n",
       "      <td>ATGCCAGCAGGTCAGTGACCTGGCTGTGAATCGAGATAAAGCCGAT...</td>\n",
       "      <td>0.919931</td>\n",
       "      <td>0.906042</td>\n",
       "      <td>0.912986</td>\n",
       "      <td>9.649436</td>\n",
       "      <td>9.649436</td>\n",
       "      <td>9.649436</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>887886</td>\n",
       "      <td>888093</td>\n",
       "      <td>+</td>\n",
       "      <td>0.021949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321119</th>\n",
       "      <td>TGCCCGTTGATTTTCAGAGAAGGGGAATTAGTACAGCAGACGGGCG...</td>\n",
       "      <td>0.740166</td>\n",
       "      <td>0.790270</td>\n",
       "      <td>0.765218</td>\n",
       "      <td>1.315612</td>\n",
       "      <td>1.315612</td>\n",
       "      <td>1.315612</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3055172</td>\n",
       "      <td>3055425</td>\n",
       "      <td>+</td>\n",
       "      <td>0.094497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321120</th>\n",
       "      <td>GGCCTGGTGGCCTGCATCGCGCTGGCGCTGATTATCGCCACGCTCG...</td>\n",
       "      <td>1.458680</td>\n",
       "      <td>1.492610</td>\n",
       "      <td>1.475645</td>\n",
       "      <td>9.895783</td>\n",
       "      <td>9.895783</td>\n",
       "      <td>9.895783</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3844004</td>\n",
       "      <td>3844212</td>\n",
       "      <td>-</td>\n",
       "      <td>0.033173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321121</th>\n",
       "      <td>AAAATGAAATTGGGCAGTTGAAACCAGACGTTTCGCCCCTATTACA...</td>\n",
       "      <td>1.211114</td>\n",
       "      <td>0.487061</td>\n",
       "      <td>0.849087</td>\n",
       "      <td>2.830200</td>\n",
       "      <td>2.830200</td>\n",
       "      <td>2.830200</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1862773</td>\n",
       "      <td>1863068</td>\n",
       "      <td>-</td>\n",
       "      <td>1.314162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321122</th>\n",
       "      <td>CAATATCGCCAACTATTTTGGTCAGTGGCCGGTGGATTATAAGAGC...</td>\n",
       "      <td>0.988964</td>\n",
       "      <td>0.842428</td>\n",
       "      <td>0.915696</td>\n",
       "      <td>9.632608</td>\n",
       "      <td>9.632608</td>\n",
       "      <td>9.632608</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2900333</td>\n",
       "      <td>2900531</td>\n",
       "      <td>+</td>\n",
       "      <td>0.231364</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>321123 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 fragment  RNA_exp_1  \\\n",
       "0       CTTCCAACCCATGGCGCGTGCGTACATAAAGGTTTCGGACGCGCGC...   1.316755   \n",
       "1       TGACTTCTTCGTGAACTTGCTGGATATGCGTTACGAGTGGAAAGCG...   0.776481   \n",
       "2       TGGGGGGATGTATGGGTACGTTGTAATTAGGGATTTAACGAATTAG...   0.530471   \n",
       "3       AAAAATAAAATTCAGCAATCAATTAATGCCTTACATCAACATGGCA...   2.392780   \n",
       "4       TTTTTTGTCGTTGACCTCACCATGTCGATCACTGTGCCTGTATCCC...   0.664520   \n",
       "...                                                   ...        ...   \n",
       "321118  ATGCCAGCAGGTCAGTGACCTGGCTGTGAATCGAGATAAAGCCGAT...   0.919931   \n",
       "321119  TGCCCGTTGATTTTCAGAGAAGGGGAATTAGTACAGCAGACGGGCG...   0.740166   \n",
       "321120  GGCCTGGTGGCCTGCATCGCGCTGGCGCTGATTATCGCCACGCTCG...   1.458680   \n",
       "321121  AAAATGAAATTGGGCAGTTGAAACCAGACGTTTCGCCCCTATTACA...   1.211114   \n",
       "321122  CAATATCGCCAACTATTTTGGTCAGTGGCCGGTGGATTATAAGAGC...   0.988964   \n",
       "\n",
       "        RNA_exp_2  RNA_exp_ave  DNA_sum_1  DNA_sum_2   DNA_ave  \\\n",
       "0        1.635498     1.476127   1.006226   1.006226  1.006226   \n",
       "1        0.881533     0.829007   2.227730   2.227730  2.227730   \n",
       "2        1.426090     0.978280   2.721586   2.721586  2.721586   \n",
       "3        2.118698     2.255739   5.572886   5.572886  5.572886   \n",
       "4        0.619367     0.641943   1.163899   1.163899  1.163899   \n",
       "...           ...          ...        ...        ...       ...   \n",
       "321118   0.906042     0.912986   9.649436   9.649436  9.649436   \n",
       "321119   0.790270     0.765218   1.315612   1.315612  1.315612   \n",
       "321120   1.492610     1.475645   9.895783   9.895783  9.895783   \n",
       "321121   0.487061     0.849087   2.830200   2.830200  2.830200   \n",
       "321122   0.842428     0.915696   9.632608   9.632608  9.632608   \n",
       "\n",
       "        num_mapped_barcodes  num_integrated_barcodes    start      end strand  \\\n",
       "0                         1                        1  1838405  1838702      -   \n",
       "1                         4                        1  4133776  4134009      +   \n",
       "2                         3                        1  2862818  2863076      -   \n",
       "3                         2                        1  3798297  3798568      -   \n",
       "4                         2                        1  3657099  3657389      +   \n",
       "...                     ...                      ...      ...      ...    ...   \n",
       "321118                    3                        1   887886   888093      +   \n",
       "321119                    1                        1  3055172  3055425      +   \n",
       "321120                    2                        1  3844004  3844212      -   \n",
       "321121                    2                        1  1862773  1863068      -   \n",
       "321122                    1                        1  2900333  2900531      +   \n",
       "\n",
       "        variation  \n",
       "0        0.312743  \n",
       "1        0.183064  \n",
       "2        1.426718  \n",
       "3        0.175510  \n",
       "4        0.101518  \n",
       "...           ...  \n",
       "321118   0.021949  \n",
       "321119   0.094497  \n",
       "321120   0.033173  \n",
       "321121   1.314162  \n",
       "321122   0.231364  \n",
       "\n",
       "[321123 rows x 13 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(os.path.join(data_path, \"frag-rLP5_LB_expression.txt\"), delimiter=\" \")\n",
    "df\n",
    "\n",
    "# Expression levels from genomic fragment MPRA (random 200–300 bp sheared fragments) in LB media\t\n",
    "    # fragment: 150–300 bp genomic DNA sequence that were randomly sheared and barcoded\n",
    "    # RNA_exp_1: RNA expression level (replicate 1) – normalized measurement of transcript abundance for this fragment in the first RNA-Seq replicate. (normalized by DNA)\n",
    "    # RNA_exp_2: RNA expression level (replicate 2) – same as above, but from a second biological replicate.\n",
    "    # RNA_exp_ave: Average RNA expression – mean of RNA_exp_1 and RNA_exp_2\n",
    "    # DNA_sum_1: DNA integration abundance (replicate 1) – quantifies how much of this DNA fragment (or barcode) was actually integrated in the first DNA-Seq sample. \n",
    "    # DNA_sum_2: DNA integration abundance (replicate 2) – same as above, from the second replicate.\n",
    "    # DNA_ave: Average DNA integration level – mean of DNA_sum_1 and DNA_sum_2\n",
    "    # num_integrated_barcodes: Number of barcodes that were integrated into the genome for this fragment. Typically should match num_mapped_barcodes unless there are sequencing/integration issues.\n",
    "    # start: Start coordinate (in bp) in the E. coli MG1655 reference genome (U00096.2). \n",
    "    # end: End coordinate in the E. coli MG1655 reference genome (U00096.2). \n",
    "    # strand: DNA strand orientation (+ or -) \n",
    "    # variation: Standard deviation or variability in RNA measurement across barcode replicates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87ab79e1-ab65-43f4-9278-ad6fb3b00361",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ce856ae-3aa8-4f03-a52b-9a8918d8e291",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = pd.read_csv(os.path.join(data_path, \"frag-rLP5-M9_expression.txt\"), delimiter=\" \")\n",
    "df2\n",
    "# Expression levels from genomic fragment MPRA (random 200–300 bp sheared fragments) in M9 media at rLP5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73fdb186-fa29-4029-8968-9cbefb0da700",
   "metadata": {},
   "outputs": [],
   "source": [
    "df4 = pd.read_csv(os.path.join(data_path, \"endo_scramble_expression_formatted_std.txt\"), delimiter= \"\\t\")\n",
    "df4\n",
    "\n",
    "# Positive promoter controls are synthetic thus have no actual location data. This refers to the 2,000 out of the 17,000 TSS they found, scrambled them to find the functional sites, including controls, scrambled variants, and unscrambled\n",
    "    # name: ID represented as {[TSS name][genomic position][strand + the scrambled region]}\n",
    "    # tss_name: Name of the original (unscrambled) transcription start site (TSS) this variant is derived from.\n",
    "    # tss_position: Start genome coordinate of the TSS.\n",
    "    # strand: Strand direction\n",
    "    # scramble_start: Start of the 10 bp scrambled window relative to var_left and var_right\n",
    "    # scramble_end: End of the scrambled window relative to var_left and var_right\n",
    "    # var_left: Genomic coordinate of the start (5' end) of the full 150 bp promoter variant.\n",
    "    # var_right: Genomic coordinate of the end (3' end) of the full 150 bp promoter variant.\n",
    "    # scramble_start_pos: Genomic coordinate where the scrambled 10 bp region begins.\n",
    "    # scramble_end_pos: Genomic coordinate where the scrambled 10 bp region ends.\n",
    "    # scramble_pos_rel_tss: Position of the scrambled region relative to the TSS.\n",
    "    # variant: The sequence\n",
    "    # RNA_exp_sum_1_1: Sum of RNA expression counts for barcodes in replicate 1, technical replicate 1 (normalized by DNA?)\n",
    "    # RNA_exp_sum_1_2: Sum of RNA expression counts for barcodes in replicate 1, technical replicate 2.\n",
    "    # RNA_exp_sum_2_1: Sum of RNA expression counts for barcodes in replicate 2, technical replicate 1.\n",
    "    # RNA_exp_sum_2_2: Sum of RNA expression counts for barcodes in replicate 2, technical replicate 2.\n",
    "    # RNA_exp_sum_1: Sum of RNA expression across both technical replicates for biological replicate 1.\n",
    "    # RNA_exp_sum_2: Sum of RNA expression across both technical replicates for biological replicate 2.\n",
    "    # RNA_exp_sum_ave: (RNA_exp_sum_1 + RNA_exp_sum_2)/2)\n",
    "    # DNA_1: DNA-seq count in biological replicate 1 (how much of this fragment was integrated).\n",
    "    # DNA_2: DNA-seq count in biological replicate 2\n",
    "    # DNA_ave: (DNA_1 + DNA_2) / 2\n",
    "    # expn_med: Median RNA expression across all barcodes, normalized by DNA abundance (RNA/DNA).\n",
    "    # num_barcodes_integrated: Number of barcodes actually integrated and measured in DNA/RNA-seq.\n",
    "    # category: Label indicating whether this is scrambled, unscrambled, negative, positive control\n",
    "    # unscrambled_exp: Measured promoter activity of the wild-type (unscrambled) version of this promoter.\n",
    "    # relative_exp: (expn_med / unscrambled_exp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9976ee5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract the content of row 1 variant column\n",
    "variant_sequence = df4.iloc[0]['variant']\n",
    "variant_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a06ad182",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(variant_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d391cd4e-cbcb-422e-a861-ad204456ac0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df5 = pd.read_csv(os.path.join(data_path, \"fLP3_Endo2_lb_expression_formatted_std.txt\"), delimiter=\"\\t\")\n",
    "df5\n",
    "# Expression summary for the TSS promoter library (17,635 reported TSSs) in LB, integrated at the fLP3 site\n",
    "    # name: Unique identifier for the promoter variant in the format [TSS_name,TSS_position,strand]\n",
    "    # tss_name: The named identifier for the transcription start site (TSS), such as from RegulonDB or Storz \n",
    "    # tss_position: Genomic coordinate of the TSS — the position where transcription starts.\n",
    "    # strand: Strand direction\n",
    "    # start: Absolute genomic coordinate where the 150 bp promoter fragment starts.\n",
    "    # end: Absolute genomic coordinate where the 150 bp promoter fragment ends.\n",
    "    # variant: The promoter sequence\n",
    "    # RNA_exp_sum_1_1: Sum of RNA-seq counts for this variant in biological replicate 1, technical replicate 1.\n",
    "    # RNA_exp_sum_1_2: Sum of RNA-seq counts in biological replicate 1, technical replicate 2.\n",
    "    # RNA_exp_sum_2_1: Sum of RNA-seq counts in biological replicate 2, technical replicate 1.\n",
    "    # RNA_exp_sum_2_2: Sum of RNA-seq counts in biological replicate 2, technical replicate 2.\n",
    "    # RNA_exp_sum_1: Total RNA expression for biological replicate 1 \n",
    "    # RNA_exp_sum_2: Total RNA expression for biological replicate 2 \n",
    "    # RNA_exp_sum_ave: Average of RNA expression across the two biological replicates.\n",
    "    # DNA_1: DNA-seq count in biological replicate 1 \n",
    "    # DNA_2: DNA-seq count in biological replicate 2.\n",
    "    # DNA_ave: Average of DNA_1 and DNA_2\n",
    "    # expn_med: Median RNA/DNA expression across barcodes \n",
    "    # num_barcodes_integrated: Number of barcodes that were actually integrated and sequenced.\n",
    "    # category: Type of promoter: ['tss', 'neg_control', 'pos_control']\n",
    "    # active: Binary classification — \"active\" or \"inactive\" promoter based on expn_med\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43d93c2d-08b9-4e49-9497-3063d42a5d52",
   "metadata": {},
   "outputs": [],
   "source": [
    "df6 = pd.read_csv(os.path.join(data_path,\"peak_tile_expression_formatted_std.txt\"), delimiter=\"\\t\")\n",
    "df6[\"category\"].unique()\n",
    "# The authors first used sheared genomic fragments (~200–300 bp) to find candidate promoter regions, and then designed a tiling oligo library: ~150 bp sequences overlapping by 10 bp Spanning all ~3,500 candidate promoter region, located at LB and nth-ydgR intergenic locus \n",
    "    # variant: The 150 bp DNA sequence (oligo) used in the MPRA tile\n",
    "    # name: Identifier for the tile, usually formatted as “peak_start_peak_end_strand_posStart-posEnd”.\n",
    "    # peak_start: Genomic start coordinate of the candidate promoter region (the whole peak region).\n",
    "    # peak_end: Genomic end coordinate of the candidate promoter region.\n",
    "    # strand: Strand direction\n",
    "    # tile_start: Start of this 150 bp oligo relative to the peak\n",
    "    # tile_end: End of the tile \n",
    "    # RNA_exp_sum_1_1: Sum of RNA reads for this tile in biological replicate 1, technical replicate 1\n",
    "    # RNA_exp_sum_1_2: Same as above, but technical replicate 2\n",
    "    # RNA_exp_sum_2_1: Sum of RNA reads in biological replicate 2, technical replicate 1.\n",
    "    # RNA_exp_sum_2_2: Same as above, but technical replicate 2.\n",
    "    # RNA_exp_sum_1: Total RNA expression in biological replicate 1\n",
    "    # RNA_exp_sum_2: Total RNA expression in biological replicate 2 \n",
    "    # RNA_exp_sum_ave: Average RNA expression across the two biological replicates.\n",
    "    # DNA_1: DNA read count in biological replicate 1\n",
    "    # DNA_2: DNA read count in biological replicate 2.\n",
    "    # DNA_ave: Average of DNA_1 and DNA_2\n",
    "    # expn_med: Median RNA/DNA expression across all barcodes for this tile \n",
    "    # num_barcodes_integrated: Number of barcodes that were successfully integrated and sequenced.\n",
    "    # category: ['tile', 'neg_control', 'random', 'pos_control']\n",
    "    # peak_length: Length (in bp) of the peak region from which this tile was derived (e.g., 362 bp).\n",
    "    # tile_start_relative: Position of the tile’s start within the peak (tile_start / peak_length)\n",
    "    # start: Genomic start coordinate of the tile (based on tile_start + peak_start).\n",
    "    # end: Genomic end coordinate of the tile.\n",
    "    # active: Binary label indicating if the tile is active or inactive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d80c50c-35eb-4be1-8bd4-a249cf1eaa3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df7 = pd.read_csv(os.path.join(data_path,\"rLP5_Endo2_lb_expression_formatted_std.txt\"), delimiter=\"\\t\")\n",
    "df7\n",
    "# Expression summary for the TSS promoter library (17,635 reported TSSs) in LB, integrated at the rLP5 site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "932b330d-c7e7-4308-9762-93fa7257ddb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df8= pd.read_csv(os.path.join(data_path, \"rLP6_Endo2_lb_expression_formatted_std.txt\"), delimiter=\"\\t\")\n",
    "df8\n",
    "# Expression summary for the TSS promoter library (17,635 reported TSSs) in LB, integrated at the rLP6 site"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "761f55c1-a68b-485f-826f-09a11af50b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "def loop_over_df(df, media=\"LB\"):\n",
    "    doc = sbol2.Document()\n",
    "    sbol2.setHomespace('http://github.com/cywlol/promoters')\n",
    "    version = '1.0'\n",
    "\n",
    "    # create the expression data so we can add it as an attachment\n",
    "    rna_df = df[[\"RNA_exp_1\", \"RNA_exp_2\", \"RNA_exp_ave\"]]\n",
    "    rna_df.loc[:, 'sample_id'] = [f'sample_{i}' for i in rna_df.index]\n",
    "    rna_df.to_csv(f'frag_{media}_expression_data.csv', index=False)\n",
    "\n",
    "    media_label_MD = media\n",
    "    chassis_label_MD = \"E_coli_chassis\"\n",
    "    strain_label_CD = \"E_coli_strain\"\n",
    "    exp_label = \"experiment_data\"\n",
    "\n",
    "    # define the experiment data once, then have it as a reference for each promoter row\n",
    "    exp_label = \"experiment_data\"\n",
    "    exp_attachment = sbol2.Attachment(exp_label)\n",
    "    exp_attachment.name = 'fragmentation expression at rLP5'\n",
    "    exp_attachment.description = 'CSV including the gene expression of the sequence: RNA_exp1, RNA_exp2, and its average.'\n",
    "    exp_attachment.source = 'CSV_LINK_HERE' # update when added attachment to SBOL collection\n",
    "    exp_attachment.format = 'https://identifiers.org/edam/format_3752'\n",
    "    doc.addAttachment(exp_attachment)\n",
    "\n",
    "    # define the media once, then have it as a reference for each promoter row\n",
    "    media_md = sbol2.ModuleDefinition(media_label_MD)\n",
    "    doc.addModuleDefinition(media_md)\n",
    "    media_md.addRole(\"http://identifiers.org/ncit/NCIT:C48164\") \n",
    "    \n",
    "    # define the chassis once, then have it as a reference for each promoter row\n",
    "    chassis_md = sbol2.ModuleDefinition(chassis_label_MD) \n",
    "    doc.addModuleDefinition(chassis_md)\n",
    "    chassis_md.addRole(\"http://identifiers.org/ncit/NCIT:C14419\")\n",
    "    #attach genome \n",
    "    attachment_label = \"ecoli_mg1655_fasta\"\n",
    "    fasta_attachment = sbol2.Attachment(attachment_label)\n",
    "    fasta_attachment.name = 'E. coli MG1655 genome.'\n",
    "    fasta_attachment.description = 'E. coli MG1655 genome sequence'\n",
    "    fasta_attachment.source = 'https:/0www.ncbi.nlm.nih.gov/nuccore/U00096.3?report=fasta'\n",
    "    fasta_attachment.format = 'https://identifiers.org/edam/format_1929'\n",
    "    chassis_md.attachments = [fasta_attachment.persistentIdentity]\n",
    "    doc.addAttachment(fasta_attachment)\n",
    "    # add tax id as a  was_derived_from relation to https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=511145\n",
    "\n",
    "    for i, row in df.iterrows():\n",
    "        fragment_seq = row[\"fragment\"]\n",
    "\n",
    "        if (i == 3):\n",
    "            break\n",
    "\n",
    "        # Establish the identities of each component, ensuring that each one is unique\n",
    "        promoter_label_CD = f\"promoter_{i}\"\n",
    "        promoter_seq_label = f\"promoter_seq_{i}\"\n",
    "        sample_design_label_CD = f\"sample_design_{i}\"\n",
    "        strain_label_CD = f\"strain_{i}\"\n",
    "        data_label = f\"exp_data_{i}\"\n",
    "        #engr_region_label_CD = f\"engr_region_{i}\"\n",
    "        #location_promoter_annotation = f\"location_promoter_annotation_{i}\"\n",
    "        #location_promoter_label = f\"location_promoter_label_{i}\"\n",
    "        \n",
    "        \n",
    "        \n",
    "        # Promoter and Sequence \n",
    "        promoter_cd = sbol2.ComponentDefinition(promoter_label_CD, sbol2.BIOPAX_DNA)\n",
    "        promoter_cd.roles = [sbol2.SO_PROMOTER]\n",
    "        seq = sbol2.Sequence(promoter_seq_label, fragment_seq, sbol2.SBOL_ENCODING_IUPAC)\n",
    "        doc.addSequence(seq)\n",
    "        promoter_cd.sequences = [seq.persistentIdentity]\n",
    "        doc.addComponentDefinition(promoter_cd)\n",
    "\n",
    "            # define the strain once, then have it as a reference for each promoter row\n",
    "        strain_md = sbol2.ModuleDefinition(strain_label_CD)\n",
    "        strain_md.addRole(\"http://identifiers.org/ncit/NCIT:C14419\")\n",
    "        doc.addModuleDefinition(strain_md)\n",
    "        strain_c1 = strain_md.modules.create('chassis')\n",
    "        strain_c1.definition = chassis_md.persistentIdentity\n",
    "        strain_c2 = strain_md.functionalComponents.create('promoter')\n",
    "        strain_c2.definition = promoter_cd.persistentIdentity\n",
    "    \n",
    "        #annotation = sbol2.SequenceAnnotation(\"promoter_location\")\n",
    "        #range = sbol2.Range(\"prange\", start, end)\n",
    "    \n",
    "        #if (strand == \"-\"):\n",
    "        #    range.orientation = sbol2.SBOL_ORIENTATION_REVERSE_COMPLEMENT\n",
    "                \n",
    "        #annotation.locations.add(range)\n",
    "        #promoter_cd.sequenceAnnotations.add(annotation)\n",
    "        '''\n",
    "        # Engineered Region \n",
    "        engineered_cd = sbol2.ComponentDefinition(engr_region_label_CD, sbol2.BIOPAX_DNA)\n",
    "        engineered_cd.roles = [\"https://identifiers.org/so/SO:0000804\"]\n",
    "        sub = engineered_cd.components.create('promoter')\n",
    "        sub.definition = promoter_cd.persistentIdentity\n",
    "        doc.addComponentDefinition(engineered_cd)\n",
    "        # No sequence?\n",
    "        '''\n",
    "        \n",
    "        #strain_c2 = strain_md.functionalComponents.create('engineered_region')\n",
    "        #strain_c2.definition = strain_md.persistentIdentity\n",
    "        \n",
    "        # Sample Design  \n",
    "        sample_md = sbol2.ModuleDefinition(sample_design_label_CD)\n",
    "        doc.addModuleDefinition(sample_md)\n",
    "        sample_md.addRole(\"http://identifiers.org/obo/OBI:0000073\")\n",
    "\n",
    "        m_strain = sample_md.modules.create('strain')\n",
    "        m_strain.definition = strain_md.persistentIdentity\n",
    "        \n",
    "        m_media = sample_md.modules.create('media')\n",
    "        m_media.definition = media_md.persistentIdentity\n",
    "    \n",
    "        # Experiment and Measurement Data\n",
    "        exp = sbol2.ExperimentalData(f'sample_{i}_expression_data')\n",
    "        exp.wasDerivedFrom.append(sample_md.persistentIdentity)\n",
    "        exp.attachments = [exp_attachment.persistentIdentity]\n",
    "        doc.add(exp)\n",
    "    doc.write('test.xml')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0de59a6b-c443-4788-a8f0-0d1235fab7e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/gonzalovidal/opt/anaconda3/envs/GLLDB/lib/python3.9/site-packages/pandas/core/indexing.py:1667: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.obj[key] = value\n"
     ]
    }
   ],
   "source": [
    "loop_over_df(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e38b79-a7b5-4d86-84bc-979c19f3c19e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "GLLDB",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
